

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>What is MISO? &mdash; MISO v1.0 documentation</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="MISO v1.0 documentation" href="#" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="#">MISO v1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <a class="reference internal image-reference" href="_images/MISO-docs-logo-small.png"><img alt="A pipeline of RNA-Seq samples as bowls of MISO soup (graphic by Lou Eisenman)" class="align-right" src="_images/MISO-docs-logo-small.png" style="width: 650.0px; height: 531.05px;" /></a>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#what-is-miso" id="id36">What is MISO?</a></li>
<li><a class="reference internal" href="#how-miso-works" id="id37">How MISO works</a></li>
<li><a class="reference internal" href="#features" id="id38">Features</a></li>
<li><a class="reference internal" href="#mailing-list" id="id39">Mailing list</a></li>
<li><a class="reference internal" href="#updates" id="id40">Updates</a></li>
<li><a class="reference internal" href="#installation" id="id41">Installation</a><ul>
<li><a class="reference internal" href="#releases" id="id42">Releases</a></li>
<li><a class="reference internal" href="#latest-version-from-github" id="id43">Latest version from GitHub</a></li>
<li><a class="reference internal" href="#installation-requirements" id="id44">Installation requirements</a></li>
<li><a class="reference internal" href="#quickstart" id="id45">Quickstart</a></li>
<li><a class="reference internal" href="#installing-the-fastmiso-version-of-miso" id="id46">Installing the <tt class="docutils literal"><span class="pre">fastmiso</span></tt> version of MISO</a><ul>
<li><a class="reference internal" href="#installing-using-a-stable-release" id="id47">Installing using a stable release</a></li>
<li><a class="reference internal" href="#installing-the-latest-version-from-github-repository" id="id48">Installing the latest version from GitHub repository</a></li>
</ul>
</li>
<li><a class="reference internal" href="#testing-the-installation" id="id49">Testing the installation</a></li>
<li><a class="reference internal" href="#testing-miso" id="id50">Testing MISO</a></li>
</ul>
</li>
<li><a class="reference internal" href="#overview" id="id51">Overview</a></li>
<li><a class="reference internal" href="#ways-of-running-miso-and-associated-file-formats" id="id52">Ways of running MISO and associated file formats</a></li>
<li><a class="reference internal" href="#using-miso-on-a-cluster" id="id53">Using MISO on a cluster</a></li>
<li><a class="reference internal" href="#alternative-event-annotations" id="id54">Alternative event annotations</a><ul>
<li><a class="reference internal" href="#human-mouse-fly-alternative-event-annotations-files" id="id55">Human/mouse/fly alternative event annotations files</a></li>
<li><a class="reference internal" href="#gff-based-alternative-events-format" id="id56">GFF-based alternative events format</a></li>
<li><a class="reference internal" href="#mapping-of-alternative-events-to-genes" id="id57">Mapping of alternative events to genes</a></li>
<li><a class="reference internal" href="#using-gff-event-annotations-with-paired-end-reads" id="id58">Using GFF event annotations with paired-end reads</a></li>
<li><a class="reference internal" href="#human-mouse-gene-models-for-isoform-centric-analyses" id="id59">Human/mouse gene models for isoform-centric analyses</a></li>
</ul>
</li>
<li><a class="reference internal" href="#running-miso" id="id60">Running MISO</a><ul>
<li><a class="reference internal" href="#invoking-miso-from-the-command-line" id="id61">Invoking MISO from the command line</a></li>
<li><a class="reference internal" href="#preparing-the-rna-seq-reads" id="id62">Preparing the RNA-Seq reads</a></li>
<li><a class="reference internal" href="#preparing-the-alternative-isoforms-annotation" id="id63">Preparing the alternative isoforms annotation</a></li>
<li><a class="reference internal" href="#configuring-miso" id="id64">Configuring MISO</a></li>
<li><a class="reference internal" href="#computing-expression-estimates-psi-psi-values" id="id65">Computing expression estimates (&#8220;Psi&#8221; / Ψ values)</a><ul>
<li><a class="reference internal" href="#using-single-end-reads" id="id66">Using single-end reads</a></li>
<li><a class="reference internal" href="#using-paired-end-reads" id="id67">Using paired-end reads</a><ul>
<li><a class="reference internal" href="#computing-the-insert-length-distribution-and-its-statistics" id="id68">Computing the insert length distribution and its statistics</a><ul>
<li><a class="reference internal" href="#getting-a-set-of-long-constitutive-exons" id="id69">Getting a set of long constitutive exons</a></li>
<li><a class="reference internal" href="#computing-the-insert-length-distribution-given-a-constitutive-exons-file" id="id70">Computing the insert length distribution given a constitutive exons file</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#distributing-miso-jobs-to-a-cluster" id="id71">Distributing MISO jobs to a cluster</a></li>
<li><a class="reference internal" href="#raw-miso-output-format" id="id72">Raw MISO output format</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#summarizing-miso-output" id="id73">Summarizing MISO output</a><ul>
<li><a class="reference internal" href="#summary-file-output-format" id="id74">Summary file output format</a></li>
</ul>
</li>
<li><a class="reference internal" href="#detecting-differentially-expressed-isoforms" id="id75">Detecting differentially expressed isoforms</a><ul>
<li><a class="reference internal" href="#output-of-samples-comparison" id="id76">Output of samples comparison</a><ul>
<li><a class="reference internal" href="#bayes-factors-file-output-format" id="id77">Bayes factors file output format</a></li>
<li><a class="reference internal" href="#raw-delta-psi-distribution-output-format" id="id78">Raw Δ Ψ distribution output format</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#example-miso-pipeline" id="id79">Example MISO pipeline</a></li>
<li><a class="reference internal" href="#interpreting-and-filtering-miso-output" id="id80">Interpreting and filtering MISO output</a><ul>
<li><a class="reference internal" href="#using-the-read-class-counts" id="id81">Using the read class counts</a></li>
<li><a class="reference internal" href="#using-the-confidence-intervals" id="id82">Using the confidence intervals</a></li>
<li><a class="reference internal" href="#filtering-differentially-expressed-events" id="id83">Filtering differentially expressed events</a></li>
</ul>
</li>
<li><a class="reference internal" href="#visualizing-and-plotting-miso-output" id="id84">Visualizing and plotting MISO output</a></li>
<li><a class="reference internal" href="#frequently-asked-questions-faq" id="id85">Frequently Asked Questions (FAQ)</a><ul>
<li><a class="reference internal" href="#answers" id="id86">Answers</a></li>
</ul>
</li>
<li><a class="reference internal" href="#single-end-event-counts-sec-format" id="id87">Single-end Event Counts (SEC) format</a><ul>
<li><a class="reference internal" href="#sec-format-specification" id="id88">SEC format specification</a><ul>
<li><a class="reference internal" href="#skipped-exons-se-format" id="id89">Skipped exons (SE) format</a></li>
<li><a class="reference internal" href="#alternative-tandem-3-utrs-tandemutr-format" id="id90">Alternative tandem 3&#8217; UTRs (TandemUTR) format</a></li>
<li><a class="reference internal" href="#retained-intron-ri-format" id="id91">Retained intron (RI) format</a></li>
<li><a class="reference internal" href="#running-miso-on-sec-files" id="id92">Running MISO on SEC files</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#advanced-uses-of-miso" id="id93">Advanced uses of MISO</a></li>
<li><a class="reference internal" href="#acknowledgements" id="id94">Acknowledgements</a></li>
<li><a class="reference internal" href="#authors" id="id95">Authors</a></li>
<li><a class="reference internal" href="#installing-python-only-miso-version" id="id96">Installing Python-only MISO version</a></li>
<li><a class="reference internal" href="#references" id="id97">References</a><ul>
<li><a class="reference internal" href="#related-software" id="id98">Related software</a></li>
<li><a class="reference internal" href="#related-reading" id="id99">Related reading</a><ul>
<li><a class="reference internal" href="#probabilistic-models-in-computational-biology" id="id100">Probabilistic models in computational biology</a></li>
<li><a class="reference internal" href="#bayes-factors-and-bayesian-hypothesis-testing" id="id101">Bayes factors and Bayesian hypothesis testing</a></li>
<li><a class="reference internal" href="#general-reading-on-probabilistic-modeling-and-inference" id="id102">General reading on probabilistic modeling and inference</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<a class="reference external image-reference" href=":ref:`installing-fastmiso`"><img alt="fastmiso, a faster version of MISO written in C" class="align-left" src="_images/speed.jpg" style="width: 173.6px; height: 125.6px;" /></a>
<a class="reference external image-reference" href="sashimi.html"><img alt="sashimi-plot" class="align-right" src="_images/sashimi-plot.png" style="width: 219.8px; height: 127.8px;" /></a>
<a href="http://www.linkedin.com/shareArticle?mini=true&url=http://genes.mit.edu/burgelab/miso/docs/&title=MISO: Analysis of isoform expression using RNA-Seq&summary=&source=MIT" target="_new">
<img src="_images/linkedin.png" alt="linkedin share button" title="linkedin share button" />
</a>
<a target="_blank" href="http://twitter.com/share?text=Currently reading MISO documentation (http%3A%2F%2Fgenes.mit.edu%2Fburgelab%2Fmiso%2Fdocs%2F)" title="Click to share this post on Twitter">
<img src="_images/twitter.png" alt="twitter share button" title="twitter share button" />
</a><p>We recently released a version of MISO written in the C language (called <tt class="docutils literal"><span class="pre">fastmiso</span></tt>) which is <strong>60-100x faster</strong> than the original Python-only version! See <a class="reference internal" href="#installing-fastmiso"><em>Installing the fastmiso version of MISO</em></a> section for more information.</p>
<p>We also released a flexible plotting tool, <a class="reference external" href="sashimi.html">sashimi_plot</a>, for visualizing RNA-Seq data and MISO output. Check it out!</p>
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr><td>&nbsp;</td>
</tr>
</tbody>
</table>
<p><strong>Quick links:</strong></p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="#event-annotation"><em>Download alternative event annotation files (available for human, mouse and D. melanogaster)</em></a></li>
<li><a class="reference internal" href="#iso-centric"><em>Download gene models in GFF format for isoform-centric analysis (available for human, mouse)</em></a></li>
</ul>
</div></blockquote>
<div class="section" id="what-is-miso">
<h1><a class="toc-backref" href="#id36">What is MISO?</a><a class="headerlink" href="#what-is-miso" title="Permalink to this headline">¶</a></h1>
<p>MISO (Mixture-of-Isoforms) is a probabilistic framework that
quantitates the expression level of alternatively spliced genes from
RNA-Seq data, and identifies differentially regulated isoforms or
exons across samples. By modeling the generative process by which
reads are produced from isoforms in RNA-Seq, the MISO model uses
Bayesian inference to compute the probability that a read originated
from a particular isoform.</p>
<p>The MISO framework is described in Katz et. al., <a class="reference external" href="http://www.nature.com/nmeth/journal/v7/n12/full/nmeth.1528.html">Analysis and design of RNA sequencing experiments for identifying isoform regulation</a>. <em>Nature Methods</em> (2010).</p>
</div>
<div class="section" id="how-miso-works">
<h1><a class="toc-backref" href="#id37">How MISO works</a><a class="headerlink" href="#how-miso-works" title="Permalink to this headline">¶</a></h1>
<p>MISO treats the expression level of a set of isoforms as a random
variable and estimates a distribution over the values of this
variable. The estimation algorithm is based on sampling, and falls in
the family of techniques known as Markov Chain Monte Carlo (&#8220;MCMC&#8221;).
For details of the inference procedure, see Katz et. al. (2010).</p>
</div>
<div class="section" id="features">
<h1><a class="toc-backref" href="#id38">Features</a><a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li>Estimates of isoform expression (Ψ values, for &#8220;Percent Spliced In&#8221; or &#8220;Percent Spliced Isoform&#8221;) and differential
isoform expression for single-end or paired-end RNA-Seq data</li>
<li>Expression estimates at the alternative splicing event level
(&#8220;exon-centric&#8221; analysis) or at the whole mRNA isoform-level (&#8220;isoform-centric&#8221; analysis)</li>
<li>Confidence intervals for expression estimates and quantitative
measures of differential expression (&#8220;Bayes factors&#8221;)</li>
<li>Basic functionality for use on cluster / distributed computing
system</li>
</ul>
</div>
<div class="section" id="mailing-list">
<h1><a class="toc-backref" href="#id39">Mailing list</a><a class="headerlink" href="#mailing-list" title="Permalink to this headline">¶</a></h1>
<p>Mailing list where users can ask technical questions is available at <a class="reference external" href="http://mailman.mit.edu/mailman/listinfo/miso-users">miso-users</a> (<a class="reference external" href="http://mailman.mit.edu/mailman/listinfo/miso-users">http://mailman.mit.edu/mailman/listinfo/miso-users</a>).</p>
</div>
<div class="section" id="updates">
<h1><a class="toc-backref" href="#id40">Updates</a><a class="headerlink" href="#updates" title="Permalink to this headline">¶</a></h1>
<p><strong>2012</strong></p>
<ul>
<li><p class="first"><strong>Tuesday, Dec 4</strong>: Posted mappings from alternative events to genes (see <a class="reference internal" href="#events-to-genes"><em>Mapping of alternative events to genes</em></a>) and GFF annotations for isoform-centric inference (see <a class="reference internal" href="#iso-centric"><em>Human/mouse gene models for isoform-centric analyses</em></a>).</p>
</li>
<li><p class="first"><strong>Thursday, Sept 27</strong>: Released <tt class="docutils literal"><span class="pre">misopy-0.4.6</span></tt>. This release fixes a packaging error with <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> (test case files were omitted.) No changes were made to MISO. Thanks to Schragi Schwartz and Rahul Satija.</p>
</li>
<li><p class="first"><strong>Tuesday, Sept 4</strong>: Released <tt class="docutils literal"><span class="pre">misopy-0.4.5</span></tt>. Fixed error in parsing of settings file.</p>
</li>
<li><p class="first"><strong>Thur, Jul 26</strong>: Released <tt class="docutils literal"><span class="pre">misopy-0.4.4</span></tt>. Turned off autoconvergence features of sampler that caused delayed run times on certain events. (Also introduced a parameter <tt class="docutils literal"><span class="pre">num_chains</span></tt> in MISO settings file to control the number of independent MCMC chains used by the sampler.)</p>
</li>
<li><p class="first"><strong>Fri, Jun 1</strong>: Released <tt class="docutils literal"><span class="pre">misopy-0.4.3</span></tt>. Included one main bug fix to an instability issue related to Bayes factors which affected low coverage events and better error handling for <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt>. The Bayes factor bug resulted in some very low coverage events having highly variable Bayes factor values across MISO runs. This issue does not affect the Ψ values or their confidence intervals, only the Bayes factor values. Events that meet most standard coverage filters applied to read counts will not be affected. Thanks to Ersen Kavak, Essi Laajala and Schragi Schwartz.</p>
</li>
<li><p class="first"><strong>Fri, May 4</strong>: Release of <tt class="docutils literal"><span class="pre">misopy-0.4.2</span></tt>. Included some bug fixes, notably:</p>
<ul class="simple">
<li>Showing complete output for events</li>
<li>Fixed bug that prevented paired-end end inserts where the read mates overlap from being aligned</li>
<li>Proper parsing of event names with multiple periods in their GFF <tt class="docutils literal"><span class="pre">ID</span></tt> field</li>
<li>Made SGE cluster submission scripts automatically executable</li>
</ul>
<p>Thanks to Rob Bradley, Eric Suh, Brad Friedman and others for their comments.</p>
</li>
<li><p class="first"><strong>Wed, Feb 1</strong>: Release of <tt class="docutils literal"><span class="pre">misopy-0.4.1</span></tt>. This is mainly a feature release. Updates include gene information being outputted in summary files and sample comparison files, dynamic scaling for <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> (described in its documentation page), options for labeling summary/comparison files, and other minor features. Bugfixes include: correct inclusion of test files with <tt class="docutils literal"><span class="pre">pypi</span></tt> releases, correction of edge cases that caused <tt class="docutils literal"><span class="pre">compare-samples</span></tt> to fail in case a multi-isoform event was improperly sampled.</p>
</li>
<li><p class="first"><strong>Fri, Jan 13</strong>: Release <tt class="docutils literal"><span class="pre">misopy-0.4</span></tt>. Includes mainly fix to serious bug that caused MISO to drop junction reads for events containing alignments with insertions (e.g. CIGAR strings containing <tt class="docutils literal"><span class="pre">I</span></tt>). Users of Tophat and other read mappers that allow insertions in alignments please take notice.</p>
</li>
<li><p class="first"><strong>Sun, Jan 8</strong>: Release <tt class="docutils literal"><span class="pre">misopy-0.3</span></tt>. Includes mainly <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> features/bugfixes.</p>
</li>
<li><p class="first"><strong>Tue, Jan 3</strong>: Released <tt class="docutils literal"><span class="pre">misopy-0.2</span></tt>. Includes bugfixes to test cases and new <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> features.</p>
</li>
</ul>
<p><strong>2011</strong></p>
<ul>
<li><p class="first"><strong>Mon, Dec 26</strong>: We&#8217;ve done some major reorganization of MISO and <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> to make them official Python packages. MISO can now be installed like as a standard Python package from the Python Package Index (see our <a class="reference external" href="http://pypi.python.org/pypi/misopy/">pypi</a> page.) The manual has been changed to reflect the new and improved installation procedure, which makes it easier to invoke MISO from the command line, requiring no modifications to your environment path variables.</p>
<blockquote>
<div><div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The new reorganized version of MISO cannot be used with older indexed/pickled files, and requires GFF files to be <strong>re-indexed</strong> by calling <tt class="docutils literal"><span class="pre">index_gff.py</span></tt> from the new package. Indexed/pickled files created with older versions of MISO will not work with the new version. We apologize for the inconvenience. Future updates to MISO will not need this re-indexing step, but our refactoring of the code results in this being required for this upgrade.</p>
</div>
</div></blockquote>
</li>
<li><p class="first"><strong>Fri, Dec 16</strong>: Added support for cluster submission using SGE &#8211; thanks to Michael Lovci for implementing this feature. Fixed edge case that caused <tt class="docutils literal"><span class="pre">filter_events</span></tt> to fail, thanks to Kuan-Ting Lin for pointing this out.</p>
</li>
<li><p class="first"><strong>Wed, Dec 14</strong>: Several new updates, including:</p>
<ul class="simple">
<li>Utilities for computing and plotting insert-length distributions. This makes use of the awesome <a class="reference external" href="http://code.google.com/p/bedtools/">bedtools</a> software written by Aaron Quinlan. Special thanks to Aaron for implementing our feature request, now available as the <tt class="docutils literal"><span class="pre">-intervals</span></tt> feature in <tt class="docutils literal"><span class="pre">tagBam</span></tt>, which we make heavy use of. Users of <a class="reference external" href="http://code.google.com/p/bedtools/">bedtools</a> should check it out!</li>
<li>Relaxed restrictions on GFF parser: can now accept &#8220;transcript&#8221; in addition to &#8220;mRNA&#8221; entries, and gracefully skip over genes with no mRNAs/transcripts in the file.</li>
<li>Posted GFF3 versions of Ensembl gene models for human/mouse to be used with isoform-centric analyses.</li>
</ul>
</li>
<li><p class="first"><strong>Sun, Dec 11</strong>: Removed platform-specific pickled files from annotation zip files.</p>
</li>
<li><p class="first"><strong>Sat, Dec 3</strong>: <a class="reference external" href="sashimi.html">sashimi_plot</a>, a tool for visualizing RNA-Seq reads along exons/junctions as well as MISO output is released!</p>
</li>
<li><p class="first"><strong>Fri, Dec 2</strong>: Underscores (<tt class="docutils literal"><span class="pre">_</span></tt>) are now allowed in GFF annotations.</p>
</li>
<li><p class="first"><strong>Fri, Nov 11 (11/11/11)</strong>: We posted a link to MISO annotations of alternative events for the Drosophila melanogaster genome, courtesy of Brent Graveley and the modENCODE project. Thanks to Brent for compiling and sharing these helpful annotations! We encourage those who work on Drosophila transcriptome datasets to try these and contribute more annotations of novel alternative RNA processing events.</p>
</li>
<li><p class="first"><strong>Thu, Oct 27</strong>: Released <tt class="docutils literal"><span class="pre">fastmiso</span></tt>, a version of MISO written in the C programming language which is 60-100x faster than the Python only version.  The MISO version contains a Python interface that is identical to the original Python-onlyMISO version, and thus will require no modification on the part of users aside from compiling the C code. An interface to MISO in the R language is also available (currently undocumented.)</p>
</li>
<li><p class="first"><strong>Tue, Oct 18</strong>: The <a class="reference external" href="http://www.mediabistro.com/fishbowldc/white-house-soup-of-the-day-64_b53593#.Tp2c76k31tA.gmail">White House adopts MISO</a>. From the article: <em>&#8220;Miso would be so upset if I didn’t tell you what it was,&#8221; said Host Chuck Todd employing one of his many puns.</em></p>
</li>
<li><p class="first"><strong>Fri, Sept 9</strong>: Human alternative event annotations (in GFF format) for hg19 are posted. Thanks to Brent Graveley for generating these.</p>
</li>
</ul>
</div>
<div class="section" id="installation">
<span id="id1"></span><h1><a class="toc-backref" href="#id41">Installation</a><a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>The primary version of MISO is called <tt class="docutils literal"><span class="pre">fastmiso</span></tt> &#8211; which is written in a combination of Python and the C programming language. This version is approximately 60-100x faster than the older Python-only version of MISO, which is no longer supported. To install <tt class="docutils literal"><span class="pre">fastmiso</span></tt>, see section <a class="reference internal" href="#installing-fastmiso"><em>Installing the fastmiso version of MISO</em></a>. (To download the Python only version of MISO, see <a class="reference internal" href="#installing-pymiso"><em>Installing Python-only MISO version</em></a>.) We recommend using <tt class="docutils literal"><span class="pre">fastmiso</span></tt>.</p>
<p>To install MISO, either download one of the stable releases (see <a class="reference internal" href="#releases"><em>Releases</em></a>) or install the latest version from our GitHub repository. See below for detailed installation instructions.</p>
<div class="section" id="releases">
<span id="id2"></span><h2><a class="toc-backref" href="#id42">Releases</a><a class="headerlink" href="#releases" title="Permalink to this headline">¶</a></h2>
<p>MISO is available as a Python package, listed as <tt class="docutils literal"><span class="pre">misopy</span></tt> in <a class="reference external" href="http://pypi.python.org/pypi/misopy/">pypi</a> (Python Package Index).</p>
<ul class="simple">
<li>MISO version 0.4.6 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.4.6.tar.gz">misopy-0.4.6.tar.gz</a>), September 27, 2012 release (<strong>Latest version</strong>)</li>
<li>MISO version 0.4.5 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.4.5.tar.gz">misopy-0.4.5.tar.gz</a>), September 4, 2012 release</li>
<li>MISO version 0.4.4 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.4.4.tar.gz">misopy-0.4.4.tar.gz</a>), July 26, 2012 release</li>
<li>MISO version 0.4.3 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.4.3.tar.gz">misopy-0.4.3.tar.gz</a>), June 1, 2012 release</li>
<li>MISO version 0.4.2 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.4.2.tar.gz">misopy-0.4.2.tar.gz</a>), May 4, 2012 release</li>
<li>MISO version 0.4.1 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.4.1.tar.gz">misopy-0.4.1.tar.gz</a>), February 1, 2012 release</li>
<li>MISO version 0.4 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.4.tar.gz">misopy-0.4.tar.gz</a>), January 13, 2012 release</li>
<li>MISO version 0.3 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.3.tar.gz">misopy-0.3.tar.gz</a>), January 8, 2012 release</li>
<li>MISO version 0.2 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.2.tar.gz">misopy-0.2.tar.gz</a>), January 3, 2012 release</li>
<li>MISO version 0.1 (<a class="reference external" href="http://pypi.python.org/packages/source/m/misopy/misopy-0.1.tar.gz">misopy-0.1.tar.gz</a>), Christmas release</li>
</ul>
</div>
<div class="section" id="latest-version-from-github">
<h2><a class="toc-backref" href="#id43">Latest version from GitHub</a><a class="headerlink" href="#latest-version-from-github" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Download latest MISO version from <tt class="docutils literal"><span class="pre">fastmiso</span></tt> GitHub <a class="reference external" href="https://github.com/yarden/MISO">repository</a> (.zip): <a class="reference external" href="https://github.com/yarden/MISO/zipball/fastmiso">https://github.com/yarden/MISO/zipball/fastmiso</a></li>
</ul>
</div>
<div class="section" id="installation-requirements">
<h2><a class="toc-backref" href="#id44">Installation requirements</a><a class="headerlink" href="#installation-requirements" title="Permalink to this headline">¶</a></h2>
<p>MISO requires a small number of Python modules and the <a class="reference external" href="http://samtools.sourceforge.net/">samtools</a> program for accessing SAM/BAM file. The requirements are:</p>
<ol class="arabic simple">
<li><a class="reference external" href="http://www.python.org">Python 2.6</a> or higher</li>
<li><a class="reference external" href="http://www.numpy.org">numpy</a> / <a class="reference external" href="http://www.scipy.org">scipy</a>. <strong>Note</strong>: MISO requires numpy version &gt; 1.4!</li>
<li><a class="reference external" href="http://matplotlib.sourceforge.net/">matplotlib</a></li>
<li><a class="reference external" href="http://samtools.sourceforge.net/">samtools</a> for accessing SAM/BAM files</li>
<li><a class="reference external" href="http://code.google.com/p/pysam/">pysam</a>, a Python library for working with SAM/BAM files through <tt class="docutils literal"><span class="pre">samtools</span></tt> (<strong>Note</strong>: MISO requires pysam version 0.6 or higher)</li>
</ol>
</div>
<div class="section" id="quickstart">
<h2><a class="toc-backref" href="#id45">Quickstart</a><a class="headerlink" href="#quickstart" title="Permalink to this headline">¶</a></h2>
<p>To quickly start using MISO, follow these steps (each of which is described in detail in the rest of the manual):</p>
<ol class="arabic">
<li><p class="first">Choose your annotation set in GFF (see <a class="reference internal" href="#event-annotation"><em>Alternative event annotations</em></a> for our own annotations, available for human, mouse and fruit fly genomes.)</p>
</li>
<li><p class="first">Index the annotation using <tt class="docutils literal"><span class="pre">index_gff.py</span></tt>:</p>
<div class="highlight-python"><pre>python index_gff.py --index SE.gff3 indexed_SE_events/</pre>
</div>
</li>
</ol>
<p>where <tt class="docutils literal"><span class="pre">SE.gff3</span></tt> is a GFF file containing descriptions of isoforms/alternative splicing events to be quantitated (e.g. skipped exons).</p>
<ol class="arabic simple" start="3">
<li>Run MISO:</li>
</ol>
<blockquote>
<div><ul>
<li><p class="first">Use <tt class="docutils literal"><span class="pre">run_events_analysis.py</span></tt> to get isoform expression estimates, optionally using a cluster to run jobs in parallel, e.g.:</p>
<div class="highlight-python"><pre>python run_events_analysis.py --compute-genes-psi indexed_SE_events/ my_sample1.bam --output-dir my_output1/ --read-len 36 --use-cluster
python run_events_analysis.py --compute-genes-psi indexed_SE_events/ my_sample2.bam --output-dir my_output2/ --read-len 36 --use-cluster</pre>
</div>
</li>
</ul>
<blockquote>
<div>where <tt class="docutils literal"><span class="pre">indexed_SE_events</span></tt> is a directory containing the indexed skipped exon events.</div></blockquote>
<ul>
<li><p class="first">Summarize MISO inferences using <tt class="docutils literal"><span class="pre">run_miso.py</span> <span class="pre">--summarize-samples</span></tt>:</p>
<div class="highlight-python"><pre>python run_miso.py --summarize-samples my_output1/ summaries/
python run_miso.py --summarize-samples my_output2/ summaries/</pre>
</div>
</li>
</ul>
</div></blockquote>
<ol class="arabic" start="4">
<li><p class="first">Make pairwise comparisons between samples to detect differentially expressed isoforms/events with <tt class="docutils literal"><span class="pre">run_miso.py</span> <span class="pre">--compare-samples</span></tt>:</p>
<div class="highlight-python"><pre>python run_miso.py --compare-samples my_output1/ my_output2/</pre>
</div>
</li>
<li><p class="first">Parse results, filtering significant events with high coverage.</p>
</li>
</ol>
<p>For a full example of running MISO, see <a class="reference internal" href="#pipeline"><em>Example MISO pipeline</em></a>. Also see the <a class="reference internal" href="#faq"><em>Frequently Asked Questions (FAQ)</em></a> page.</p>
</div>
<div class="section" id="installing-the-fastmiso-version-of-miso">
<span id="installing-fastmiso"></span><h2><a class="toc-backref" href="#id46">Installing the <tt class="docutils literal"><span class="pre">fastmiso</span></tt> version of MISO</a><a class="headerlink" href="#installing-the-fastmiso-version-of-miso" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">fastmiso</span></tt> version of MISO is written in a combination of C and Python. The underlying statistical inference engine is written in C and can be compiled on Mac OS X or Unix-like platforms using GCC or other standard compilers. The interface to this code is written in Python and is identical in its commands and input and output formats to the original Python-only version of MISO. (The Python-only version of MISO is now deprecated.)</p>
<p>There are two options for installing MISO, using either a stable release or installing the latest version from GitHub.</p>
<div class="section" id="installing-using-a-stable-release">
<h3><a class="toc-backref" href="#id47">Installing using a stable release</a><a class="headerlink" href="#installing-using-a-stable-release" title="Permalink to this headline">¶</a></h3>
<p>To install a stable release package of MISO using a Python package manager like <tt class="docutils literal"><span class="pre">easy_install</span></tt>, simply type:</p>
<div class="highlight-python"><pre>easy_install misopy</pre>
</div>
<p>This will install MISO globally. For local installation, use:</p>
<div class="highlight-python"><pre>easy_install --user -U misopy</pre>
</div>
<p>If <tt class="docutils literal"><span class="pre">easy_install</span></tt> is unavailable on your system, you can simply download a stable release package (available from <a class="reference internal" href="#releases"><em>Releases</em></a>) and install it
like ordinary Python modules. Unzip the package (e.g. for version &#8220;0.x&#8221;):</p>
<div class="highlight-python"><pre>tar -xzvf misopy-0.x.tar.gz</pre>
</div>
<p>This will create a <tt class="docutils literal"><span class="pre">setup.py</span></tt> file in the directory where the package was unzipped. Compile the code and install MISO packages as follows:</p>
<div class="highlight-python"><pre>python setup.py install</pre>
</div>
<p>This command will install MISO globally on the system. To install it in a particular location, such as your home directory, use the following call to <tt class="docutils literal"><span class="pre">setup.py</span></tt> instead:</p>
<div class="highlight-python"><pre>python setup.py install --prefix=~/</pre>
</div>
<p>This will result in a local installation of the MISO Python package. To test the installation, load Python and import the <tt class="docutils literal"><span class="pre">misopy</span></tt> and <tt class="docutils literal"><span class="pre">pysplicing</span></tt> packages in the interpreter:</p>
<div class="highlight-python"><pre>&gt;&gt; import misopy
&gt;&gt; import pysplicing</pre>
</div>
<p>If you get no errors, the installation completed successfully.</p>
</div>
<div class="section" id="installing-the-latest-version-from-github-repository">
<h3><a class="toc-backref" href="#id48">Installing the latest version from GitHub repository</a><a class="headerlink" href="#installing-the-latest-version-from-github-repository" title="Permalink to this headline">¶</a></h3>
<p>To install the latest version from the GitHub <a class="reference external" href="https://github.com/yarden/MISO">repository</a>, there are two options: (a) using <tt class="docutils literal"><span class="pre">git</span></tt> to get the latest source from our GitHub repository, or (b) by downloading a snapshot of the current repository, without going through <tt class="docutils literal"><span class="pre">git</span></tt>. If you have <tt class="docutils literal"><span class="pre">git</span></tt> installed and configured with your GitHub account, you can clone our repository as follows:</p>
<div class="highlight-python"><pre>git clone git://github.com/yarden/MISO.git</pre>
</div>
<p>This will create a directory called <tt class="docutils literal"><span class="pre">MISO</span></tt>, containing the repository. Alternatively, if you do not want to go through <tt class="docutils literal"><span class="pre">git</span></tt>, you can download a zip file containing the latest MISO GitHub release (following this link: <a class="reference external" href="https://github.com/yarden/MISO/zipball/fastmiso">https://github.com/yarden/MISO/zipball/fastmiso</a>). This zip file has to be unzipped (e.g. using <tt class="docutils literal"><span class="pre">unzip</span> <span class="pre">fastmiso</span></tt> on Unix systems) and contains the latest release of the MISO repository.</p>
<p>Once you&#8217;ve obtained the MISO repository, either through <tt class="docutils literal"><span class="pre">git</span></tt> or by downloading the zip file snapshot, enter the directory and prepare the C code for compilation using <tt class="docutils literal"><span class="pre">make</span></tt>:</p>
<div class="highlight-python"><pre>make Pythonpackage</pre>
</div>
<p>Next, compile the code and install MISO using <tt class="docutils literal"><span class="pre">setup.py</span></tt> as described above:</p>
<div class="highlight-python"><pre>python setup.py install</pre>
</div>
<p>The <tt class="docutils literal"><span class="pre">--prefix=</span></tt> option can be used as usual to specify an alternative installation path. For example, to install MISO in your local directory <tt class="docutils literal"><span class="pre">/home/user/pylibs/</span></tt>, you can use:</p>
<div class="highlight-python"><pre>python setup.py install --prefix=/home/user/pylibs/</pre>
</div>
<p>This will create a directory called <tt class="docutils literal"><span class="pre">lib</span></tt> inside that directory that contains the <tt class="docutils literal"><span class="pre">site-packages</span></tt> directory, which is the directory Python expects to find packages in. The exact path of the directory depends on the Python version used when calling <tt class="docutils literal"><span class="pre">setup.py</span></tt>. If you used Python 2.7, your directory might be <tt class="docutils literal"><span class="pre">/home/user/pylibs/lib/python2.7/site-packages/</span></tt>. To ensure that this directory is available to Python, you can add it to your <tt class="docutils literal"><span class="pre">PYTHONPATH</span></tt> environment variable, as follows (for <tt class="docutils literal"><span class="pre">bash</span></tt>-shell like systems):</p>
<div class="highlight-python"><pre>export PYTHONPATH=/home/user/pylibs/python2.7/site-packages/:$PYTHONPATH</pre>
</div>
<p>This will place the local <tt class="docutils literal"><span class="pre">site-packages</span></tt> ahead of the current setting of <tt class="docutils literal"><span class="pre">PYTHONPATH</span></tt>, resulting in Python loading the version of the package installed there (if it is available) over versions existing in the system&#8217;s global <tt class="docutils literal"><span class="pre">site-packages</span></tt> directory.</p>
<p>Executable scripts that are part of MISO (like <tt class="docutils literal"><span class="pre">run_events_analysis.py</span></tt>) get placed in a local <tt class="docutils literal"><span class="pre">bin</span></tt> directory (e.g. <tt class="docutils literal"><span class="pre">~/bin/</span></tt>) directory. If you&#8217;d like these to get used from the shell in place of a globally installed version, make sure that <tt class="docutils literal"><span class="pre">~/bin</span></tt> is first in your <tt class="docutils literal"><span class="pre">PATH</span></tt> variable.</p>
</div>
</div>
<div class="section" id="testing-the-installation">
<span id="id3"></span><h2><a class="toc-backref" href="#id49">Testing the installation</a><a class="headerlink" href="#testing-the-installation" title="Permalink to this headline">¶</a></h2>
<p>To test if the required modules are available, use the
<tt class="docutils literal"><span class="pre">module_availability.py</span></tt> script as follows:</p>
<div class="highlight-python"><pre>$ python module_availability.py
Checking for availability of: numpy
Checking for availability of: scipy
...
All modules are available!</pre>
</div>
<p>It is crucial that the <tt class="docutils literal"><span class="pre">samtools</span></tt> command is available from the
default path environment, or else MISO will not be able to parse
SAM/BAM files. On a Unix or a Mac OS X system, you should be able to
type <tt class="docutils literal"><span class="pre">samtools</span></tt> at the prompt and get access to the program without
specifying any path prefix.</p>
<p>To test that the MISO packages are properly installed, verify that the two modules <tt class="docutils literal"><span class="pre">misopy</span></tt> and <tt class="docutils literal"><span class="pre">pysplicing</span></tt> are available from your Python environment:</p>
<div class="highlight-python"><pre>&gt;&gt; import misopy
&gt;&gt; import pysplicing</pre>
</div>
<p>You should be able to import both of these packages without errors from the Python interpreter. The <tt class="docutils literal"><span class="pre">misopy</span></tt> package is the main MISO package, containing utilities for analyzing RNA-Seq data and plotting it (including <a class="reference external" href="sashimi.html">sashimi_plot</a>). The <tt class="docutils literal"><span class="pre">pysplicing</span></tt> package is an under-the-hood interface between Python and the statistical inference engine written in C.</p>
</div>
<div class="section" id="testing-miso">
<h2><a class="toc-backref" href="#id50">Testing MISO</a><a class="headerlink" href="#testing-miso" title="Permalink to this headline">¶</a></h2>
<p>To test that MISO can be run properly, run <tt class="docutils literal"><span class="pre">test_miso.py</span></tt> as shown
below. These tests ensure that MISO can be run on a few
genes/exons. The output of these tests can be ignored and the
abbreviated version should be along the following lines:</p>
<div class="highlight-python"><pre>$ python misopy/test_miso.py
Testing conversion of SAM to BAM...
...output omitted...
Computing Psi for 1 genes...
...
----------------------------------------------------------------------
Ran 3 tests in 68.000s
OK</pre>
</div>
<p>If you see errors prior to this, the most likely cause is that some dependency
is not installed or not available in the path environment. For
example, if <tt class="docutils literal"><span class="pre">samtools</span></tt> is installed but missing from the path
(i.e. not accessible from the shell), that will cause errors in the
test run.</p>
</div>
</div>
<div class="section" id="overview">
<h1><a class="toc-backref" href="#id51">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<p>To run MISO, a set of annotations (typically specified in GFF format) of the isoforms of alternative
events must be provided and RNA-Seq reads (typically specified in SAM format). The GFF annotation is
indexed into an efficient representation using a script provided by MISO. The SAM reads must be sorted
and indexed into the BAM format (binary version of SAM) before they can be used with MISO.</p>
<p>The sections that follow describe in detail the file formats used for the annotation and for the reads
and the specific steps that are required to run MISO. An example of a
MISO pipeline for computing isoform expression estimates and detecting
differentially expressed isoforms is given in <a class="reference internal" href="#pipeline"><em>Example MISO pipeline</em></a>.</p>
<p>The figure below shows an overview of how to run MISO.</p>
<blockquote>
<div><a class="reference internal image-reference" href="_images/MISO-graphical-pipeline.png"><img alt="Overview of how to run MISO" class="align-center" src="_images/MISO-graphical-pipeline.png" style="width: 815.1px; height: 916.2px;" /></a>
</div></blockquote>
</div>
<div class="section" id="ways-of-running-miso-and-associated-file-formats">
<h1><a class="toc-backref" href="#id52">Ways of running MISO and associated file formats</a><a class="headerlink" href="#ways-of-running-miso-and-associated-file-formats" title="Permalink to this headline">¶</a></h1>
<p>MISO can be run on either single or paired-end RNA-Seq data. Two general kinds of analyses are possible:</p>
<ol class="arabic simple">
<li>Estimate expression level of exons (<strong>&#8220;exon-centric&#8221; analysis</strong>) or,</li>
<li>Estimate expression level of whole transcripts (<strong>&#8220;isoform-centric&#8221; analysis</strong>)</li>
</ol>
<p>Exon-centric analyses are recommended for looking at alternative splicing at the level of individual
splicing events, e.g. the inclusion levels of a particular skipped exon, or the use of a particular
alternative splice site. In isoform-centric analyses the expression level of whole isoforms for genes are estimated (i.e. the
expression of each individual isoform of a gene is estimated). Both of these analysis modes have merits and disadvantages. Exon-centric analyses are typically easier to interpret and validate experimentally, but do not always capture the complexity of a set related splicing events within a gene. Isoform-centric analyses capture more of this complexity, but are limited by the typically short length of RNA-Seq reads, and inaccuracies or incompleteness in the annotation of the gene&#8217;s isoforms. Estimates of isoform-level expression are also harder to validate in the lab using traditional techniques such as RT-PCR.</p>
<p>To run MISO, an annotation of the alternative splicing events or isoforms must be provided and
a set of reads. Single or paired-end reads can be provided in the <a class="reference external" href="http://samtools.sourceforge.net/SAM1.pdf">Spliced Alignment/Map (SAM)</a> format (in its binary form, BAM) and the annotation can be in the <a class="reference external" href="http://www.sequenceontology.org/gff3.shtml">GFF</a> (version 3) format. The GFF format can be used to specify either
whole mRNA isoforms of genes (which is the common use of the format) or to specify single alternative
splicing events, as described in <a class="reference internal" href="#event-annotation"><em>Alternative event annotations</em></a>.</p>
</div>
<div class="section" id="using-miso-on-a-cluster">
<h1><a class="toc-backref" href="#id53">Using MISO on a cluster</a><a class="headerlink" href="#using-miso-on-a-cluster" title="Permalink to this headline">¶</a></h1>
<p>It is highly recommended to run MISO on a cluster. Since each
gene/exon can be treated as an independent inference problem, MISO is
highly parallelizable, and the use of multiple processors can cut run
times from days to hours on most data sets.</p>
<p>MISO comes with basic functionality for running on a cluster, assuming
that there&#8217;s a shared filesystem between all the nodes.</p>
</div>
<div class="section" id="alternative-event-annotations">
<span id="event-annotation"></span><h1><a class="toc-backref" href="#id54">Alternative event annotations</a><a class="headerlink" href="#alternative-event-annotations" title="Permalink to this headline">¶</a></h1>
<p>Alternative event annotations are available for the major classes of alternative
splicing and alternative RNA processing events in the human (hg18, hg19), mouse (mm9) and Drosophila melanogaster genomes.
The event types covered are:</p>
<ol class="arabic simple">
<li>Skipped exons (SE)</li>
<li>Alternative 3&#8217;/5&#8217; splice sites (A3SS, A5SS)</li>
<li>Mutually exclusive exons (MXE)</li>
<li>Tandem 3&#8217; UTRs (TandemUTR)</li>
<li>Retained introns (RI)</li>
<li>Alternative first exons (AFE)</li>
<li>Alternative last exons (ALE)</li>
</ol>
<p>For performing isoform-centric analyses, any gene models annotation can be used (e.g. from Ensembl, UCSC or RefSeq) as long as it is specified in the GFF3 format. As an example, we provide GFF3 annotations from Ensembl (which were converted from Ensembl&#8217;s GTF format to GFF3), available in <a class="reference internal" href="#iso-centric"><em>Human/mouse gene models for isoform-centric analyses</em></a>.</p>
<div class="section" id="human-mouse-fly-alternative-event-annotations-files">
<h2><a class="toc-backref" href="#id55">Human/mouse/fly alternative event annotations files</a><a class="headerlink" href="#human-mouse-fly-alternative-event-annotations-files" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/modENCODE_alt_events.zip">Drosophila melanogaster alternative events (modENCODE)</a> (3.2M .zip)</li>
<li><a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/mm9_alt_events.zip">Mouse genome (mm9) alternative events</a> (7.9M, .zip)</li>
<li><a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/hg18_alt_events.zip">Human genome (hg18) alternative events</a> (45M, .zip)</li>
<li><a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/hg19_alt_events.zip">Human genome (hg19) alternative events</a> (8.6M, .zip)</li>
</ul>
<p>These annotations include GFF files (<tt class="docutils literal"><span class="pre">.gff3</span></tt> extension) that can be used with MISO. The annotations for human and mouse were compiled as described in Wang et. al. (2008). Briefly, each splicing event was considered alternative if it was supported by several ESTs, and alternative tandem 3&#8217; UTRs (TandemUTR events) were derived from <a class="reference external" href="http://polya.umdnj.edu/polyadb/">PolyA DB</a>.</p>
</div>
<div class="section" id="gff-based-alternative-events-format">
<span id="gff-event-annotation"></span><h2><a class="toc-backref" href="#id56">GFF-based alternative events format</a><a class="headerlink" href="#gff-based-alternative-events-format" title="Permalink to this headline">¶</a></h2>
<p>Single alternative splicing events can be described in GFF format, using the convention
described below. In the mouse and human genome event annotations in <a class="reference internal" href="#event-annotation"><em>Alternative event annotations</em></a>,
alternative splicing events that produce two isoforms are described.
For example, here is a skipped exon from the set of mouse skipped exons annotation:</p>
<div class="highlight-python"><pre>chr1  SE      gene    4772649 4775821 .       -       .       ID=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-;Name=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-
chr1  SE      mRNA    4772649 4775821 .       -       .       ID=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.A;Parent=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-
chr1  SE      mRNA    4772649 4775821 .       -       .       ID=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.B;Parent=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-
chr1  SE      exon    4775654 4775821 .       -       .       ID=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.A.up;Parent=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.A
chr1  SE      exon    4774032 4774186 .       -       .       ID=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.A.se;Parent=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.A
chr1  SE      exon    4772649 4772814 .       -       .       ID=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.A.dn;Parent=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.A
chr1  SE      exon    4775654 4775821 .       -       .       ID=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.B.up;Parent=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.B
chr1  SE      exon    4772649 4772814 .       -       .       ID=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.B.dn;Parent=chr1:4775654:4775821:-@chr1:4774032:4774186:-@chr1:4772649:4772814:-.B</pre>
</div>
<p>The ID of the exon was chosen arbitrarily to be <tt class="docutils literal"><span class="pre">chr1:4775654:4775821:-&#64;chr1:4774032:4774186:-&#64;chr1:4772649:4772814:-</span></tt>. This name encodes the genomic coordinates of the upstream (5&#8217;) exon, the skipped exon, and
the downstream (3&#8217;) exon of this alternative splicing event, separated by <tt class="docutils literal"><span class="pre">&#64;</span></tt> symbols. The two isoforms that result (encoded as <tt class="docutils literal"><span class="pre">mRNA</span></tt> entries in the file) are: (1) the isoform made up of the upstream exon, the skipped exon and the
downstream exon (isoform where the exon is included), (2) the isoform made up of the upstream
exon followed by the downstream exon (isoform where the exon is skipped.) Note that we only use
the two flanking exons in this annotation, but one could modify the annotation to include arbitrarily many
flanking exons from the transcripts. We chose to name the two <tt class="docutils literal"><span class="pre">mRNA</span></tt> entries as the coordinates of the
exons that make up the mRNA followed by <tt class="docutils literal"><span class="pre">A</span></tt> for the first isoform, <tt class="docutils literal"><span class="pre">B</span></tt> for the second, though these
naming conventions are arbitrary and do not have to be used. Similarly, the ID of each exon is encoded
as its coordinate followed by <tt class="docutils literal"><span class="pre">up</span></tt>, <tt class="docutils literal"><span class="pre">se</span></tt> or <tt class="docutils literal"><span class="pre">dn</span></tt> to denote the upstream, skipped or downstream
exons, respectively.</p>
</div>
<div class="section" id="mapping-of-alternative-events-to-genes">
<span id="events-to-genes"></span><h2><a class="toc-backref" href="#id57">Mapping of alternative events to genes</a><a class="headerlink" href="#mapping-of-alternative-events-to-genes" title="Permalink to this headline">¶</a></h2>
<p>Below is a mapping of alternative events to genes, based on Ensembl annotation. These are tab-delimited files the first column (<tt class="docutils literal"><span class="pre">event_id</span></tt>) is the <tt class="docutils literal"><span class="pre">ID</span></tt> of the event from its GFF file and the second column (<tt class="docutils literal"><span class="pre">gene_id</span></tt>) corresponds to a comma-separated list of Ensembl identifiers for the gene(s) the event overlaps. If the event overlaps multiple genes (which could happen because multiple Ensembl identifiers are sometimes given to the same gene, or because the genes overlap and/or are contained within each other in the annotation), then multiple Ensembl identifiers will be listed. A mapping file is given for each event type (e.g. skipped exons, tandem 3&#8217; UTRs, etc.) Events that cannot be mapped to genes are recorded as <tt class="docutils literal"><span class="pre">NA</span></tt>.</p>
<ul class="simple">
<li>Mapping for mouse alternative events:<ul>
<li><a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/mm9_events_to_genes.zip">Events to genes mapping for mm9</a> (.zip)</li>
</ul>
</li>
<li>Mapping for human alternative events:<ul>
<li><a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/hg18_events_to_genes.zip">Events to genes mapping for hg18</a> (.zip)</li>
<li><a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/hg19_events_to_genes.zip">Events to genes mapping for hg19</a> (.zip)</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="using-gff-event-annotations-with-paired-end-reads">
<h2><a class="toc-backref" href="#id58">Using GFF event annotations with paired-end reads</a><a class="headerlink" href="#using-gff-event-annotations-with-paired-end-reads" title="Permalink to this headline">¶</a></h2>
<p>Note that in the above GFF annotations for events, only immediately flanking exons to an alternative event are considered.
For example, for a skipped exon event, only the two exons that flank it are considered in the annotation. In the human/mouse genomes,
this means that the two isoforms produced the event would be roughly ~400 nt and ~300 nt long. Modified annotations
that use more of the exons present in the two isoforms might be more appropriate for use with long insert length paired-end
libraries where the average insert is significantly longer than these isoform lengths. This in general will not affect isoform-centric analyses, where the entire mRNA transcript is used as the annotation to quantitate.</p>
</div>
<div class="section" id="human-mouse-gene-models-for-isoform-centric-analyses">
<span id="iso-centric"></span><h2><a class="toc-backref" href="#id59">Human/mouse gene models for isoform-centric analyses</a><a class="headerlink" href="#human-mouse-gene-models-for-isoform-centric-analyses" title="Permalink to this headline">¶</a></h2>
<p>For isoform-centric analyses, annotations of gene models are needed, where whole mRNA isoforms are specified for each gene. Any annotation that is in GFF3 format can be used, e.g. annotations obtained from RefSeq, Ensembl, UCSC or other databases.</p>
<p>Annotations can be converted from UCSC table formats or from GTF format into GFF3 using the following procedure:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Download your gene annotation (e.g. Ensembl genes, RefSeq) from UCSC Table Browser. For example, an Ensembl genes table <tt class="docutils literal"><span class="pre">ensGene.txt</span></tt> can be downloaded for the <tt class="docutils literal"><span class="pre">mm9</span></tt> genome from UCSC goldenPath here (<a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/mm9/database/ensGene.txt.gz">http://hgdownload.cse.ucsc.edu/goldenPath/mm9/database/ensGene.txt.gz</a>).</li>
<li>To convert this gene table into GFF3, use of these two ways: (a) use the script <a class="reference external" href="http://genes.mit.edu/burgelab/miso/scripts/ucsc_table2gff3.pl">ucsc_table2gff3.pl</a> (from <a class="reference external" href="http://code.google.com/p/biotoolbox/">biotoolbox</a>) to convert the table directly into GFF3, or (b) use the UCSC utility <a class="reference external" href="http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/">genePredToGtf</a> to convert the table into GTF format (instructions for this are described in <a class="reference external" href="http://genomewiki.ucsc.edu/index.php/Genes_in_gtf_or_gff_format">http://genomewiki.ucsc.edu/index.php/Genes_in_gtf_or_gff_format</a>) and then convert the GTF file into GFF3 using <a class="reference external" href="http://genes.mit.edu/burgelab/miso/scripts/gtf2gff3.pl">gtf2gff3.pl</a>.</li>
</ol>
</div></blockquote>
<p>For convenience, we provide GFF3 annotations based on UCSC Table Browser&#8217;s version of Ensembl genes for the following genomes:</p>
<blockquote>
<div><ul class="simple">
<li>Mouse Ensembl genes from UCSC<ul>
<li>Annotation for mm9: <a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/ucsc_tables/mm9/ensGene.gff3">mm9 ensGene GFF annotation</a></li>
<li>Annotation for mm10: <a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/ucsc_tables/mm10/ensGene.gff3">mm10 ensGene GFF annotation</a></li>
</ul>
</li>
<li>Human Ensembl genes from UCSC<ul>
<li>Annotation for hg18: <a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/ucsc_tables/hg18/ensGene.gff3">hg18 ensGene GFF annotation</a></li>
<li>Annotation for hg19: <a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/ucsc_tables/hg19/ensGene.gff3">hg19 ensGene GFF annotation</a></li>
</ul>
</li>
</ul>
</div></blockquote>
<p>It is critical to note that Ensembl and UCSC have distinct chromosome naming conventions (see warning note below). However, Ensembl tables obtained from the UCSC Table Browser will follow the UCSC chromosome naming conventions, not Ensembl&#8217;s, which is why the UCSC-downloaded Ensembl annotations in the GFFs provided above contain <cite>chr</cite> headers. Ensembl annotations obtained directly from Ensembl (e.g. via Ensembl BioMart) will of course follow the Ensembl chromosome naming conventions, and not UCSC&#8217;s.</p>
<blockquote>
<div><div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p>The UCSC and Ensembl databases name their chromosomes differently. By convention, UCSC chromosomes start with <tt class="docutils literal"><span class="pre">chr</span></tt> while Ensembl chromosome names do not. UCSC will call the fifth chromosome <tt class="docutils literal"><span class="pre">chr5</span></tt> and Ensembl will call it <tt class="docutils literal"><span class="pre">5</span></tt>. If your GFF annotation uses one convention, but your BAM reads are aligned to distinct <em>headers</em>, i.e. chromosome names following another convention, then none of your reads will map. For example, MISO will try to find all the regions on <tt class="docutils literal"><span class="pre">chr5</span></tt> if your GFF annotation follows the UCSC convention and fail, since all your BAM alignments could be to a header named <tt class="docutils literal"><span class="pre">5</span></tt>, or vice versa. In that case, lines such as these will appear in the MISO output:</p>
<div class="highlight-python"><pre>Cannot fetch reads in region: 5:1200-1500</pre>
</div>
<p>We strongly recommend that you check that your BAM files follow the same headers naming convention as your GFF. If you run into a mismatch between your annotation&#8217;s chromosome headers and your BAM file headers, please make sure that the headers that you map to in the read mapping stage to are the same as the headers used in the GFF annotation for MISO. For example, if you use TopHat/Bowtie to map reads and your genome index was made with UCSC-style <tt class="docutils literal"><span class="pre">chr</span></tt> headers but you want to use Ensembl annotations for quantitations, the solution is to:</p>
<blockquote class="last">
<div><ol class="arabic simple">
<li>Download your genome&#8217;s sequence files (<tt class="docutils literal"><span class="pre">*.fa</span></tt>, typically) from Ensembl,</li>
<li>Make a new Bowtie index from these files (using <tt class="docutils literal"><span class="pre">bowtie-build</span></tt>), or the appropriate index for your read mapping program,</li>
<li>Remap your reads against this new index. The resulting mapped SAM/BAM files can be used with the unmodified Ensembl GFF annotations we provide here, where the headers in the GFF are guaranteed to match the headers in the SAM/BAM files (both using Ensembl header conventions.)</li>
</ol>
</div></blockquote>
</div>
</div></blockquote>
<p>For convenience, we also provide GFF3 annotations of gene models from Ensembl (release 65), which were simply converted from Ensembl&#8217;s GTF to GFF3 format and are otherwise identical to the Ensembl annotation.</p>
<blockquote>
<div><ul class="simple">
<li>Mouse Ensembl genes (16 M, .zip): <a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/gene-models/Mus_musculus.NCBIM37.65.gff.zip">Mus_musculus.NCBIM37.65.gff</a></li>
<li>Human Ensembl genes (26 M, .zip): <a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/gene-models/Homo_sapiens.GRCh37.65.gff.zip">Homo_sapiens.GRCh37.65.gff</a></li>
</ul>
</div></blockquote>
</div>
</div>
<div class="section" id="running-miso">
<h1><a class="toc-backref" href="#id60">Running MISO</a><a class="headerlink" href="#running-miso" title="Permalink to this headline">¶</a></h1>
<div class="section" id="invoking-miso-from-the-command-line">
<h2><a class="toc-backref" href="#id61">Invoking MISO from the command line</a><a class="headerlink" href="#invoking-miso-from-the-command-line" title="Permalink to this headline">¶</a></h2>
<p>When installing the MISO package, the set of scripts that make up to inteface to MISO will automatically be placed as an executables in your path, so that you can refer to these without modifying your shell environment. For example, if you install MISO using <tt class="docutils literal"><span class="pre">setup.py</span> <span class="pre">install</span></tt> globally, then a MISO script like <tt class="docutils literal"><span class="pre">run_events_analysis.py</span></tt> will become available on the path, and can be run as:</p>
<div class="highlight-python"><pre>run_events_analysis.py ...</pre>
</div>
<p>Or alternatively, as:</p>
<div class="highlight-python"><pre>python run_events_analysis.py ...</pre>
</div>
<p>Though the <tt class="docutils literal"><span class="pre">python</span></tt> prefix is not necessary. The path of the Python version on the system used to install MISO will automatically be invoked when running these scripts.</p>
<p>If you installed MISO using <tt class="docutils literal"><span class="pre">setup.py</span> <span class="pre">--prefix=~/</span></tt> in your home directory or some other path, these scripts will typically be placed in a designated executables directory in your home directory, e.g. <tt class="docutils literal"><span class="pre">~/bin/</span></tt>.</p>
</div>
<div class="section" id="preparing-the-rna-seq-reads">
<h2><a class="toc-backref" href="#id62">Preparing the RNA-Seq reads</a><a class="headerlink" href="#preparing-the-rna-seq-reads" title="Permalink to this headline">¶</a></h2>
<p>MISO can take as input RNA-Seq reads in the <a class="reference external" href="http://samtools.sourceforge.net/SAM1.pdf">SAM</a> format. SAM files are typically the output of read alignment programs, such as <a class="reference external" href="http://bowtie-bio.sourceforge.net/">Bowtie</a> or <a class="reference external" href="http://tophat.cbcb.umd.edu/">Tophat</a>. The SAM format can represent both single and paired-end reads.</p>
<p>Files in SAM format must be converted to an indexed, sorted BAM file that contain headers.
If you have an indexed, sorted BAM file you can move on to the next section.</p>
<p>The right BAM file can be created using <a class="reference external" href="http://samtools.sourceforge.net/">samtools</a> manually, or with the help of the wrapper
script <tt class="docutils literal"><span class="pre">sam_to_bam.py</span></tt> that calls samtools. For example, to convert <tt class="docutils literal"><span class="pre">my_sample.sam</span></tt> to an
indexed, sorted BAM file with headers specified in <tt class="docutils literal"><span class="pre">mm9_complete.fa.fai</span></tt> and output the result
to the directory <tt class="docutils literal"><span class="pre">my_sample/</span></tt>, use:</p>
<div class="highlight-python"><pre>python sam_to_bam.py --convert my_sample.sam my_sample/ --ref mm9_complete.fa.fai</pre>
</div>
<p>The <tt class="docutils literal"><span class="pre">mm9_complete.fa.fai</span></tt> is a header file that contains the lengths of each chromosome
in the genome used to create the <tt class="docutils literal"><span class="pre">my_sample.sam</span></tt> file. The .fai files can be created from the
FASTA files of a genome using the <tt class="docutils literal"><span class="pre">faidx</span></tt> option of samtools (e.g. <tt class="docutils literal"><span class="pre">samtools</span> <span class="pre">faidx</span> <span class="pre">genome.fasta</span></tt>).</p>
<p>The <tt class="docutils literal"><span class="pre">sam_to_bam.py</span></tt> script will output a set of BAM files and their indices (<tt class="docutils literal"><span class="pre">my_sample.bam</span></tt>,
<tt class="docutils literal"><span class="pre">my_sample.sorted.bam</span></tt>, <tt class="docutils literal"><span class="pre">my_sample.sorted.bam.bai</span></tt>) to <tt class="docutils literal"><span class="pre">my_sample/</span></tt>.</p>
</div>
<div class="section" id="preparing-the-alternative-isoforms-annotation">
<span id="indexing-annotation"></span><h2><a class="toc-backref" href="#id63">Preparing the alternative isoforms annotation</a><a class="headerlink" href="#preparing-the-alternative-isoforms-annotation" title="Permalink to this headline">¶</a></h2>
<p>To quantitate alternative splicing, an annotation file must be provided that describes
the isoforms whose expression should be estimated from data.</p>
<p>To estimate the expression level of whole mRNA transcripts, a GFF format containing a set
of annotated transcripts for each gene can be used. This is the same GFF that can be
given as input to <a class="reference external" href="http://tophat.cbcb.umd.edu/">Tophat</a> in order to incorporate known genomic junctions, for example.
These GFF files can be readily produced for any annotated genome (see <a class="reference internal" href="#iso-centric"><em>Human/mouse gene models for isoform-centric analyses</em></a> for Ensembl-based annotations of this form for human/mouse genomes.)</p>
<p>For exon-centric analyses, a GFF file can be used as well, except such a file will contain
only the subsets of the isoforms produced by a particular splicing event &#8211; rather than
whole transcripts &#8211; as described in <a class="reference internal" href="#gff-event-annotation"><em>GFF-based alternative events format</em></a>.  The GFF file chosen has to be
indexed for efficient representation. A set of indexed, ready-to-use events is available for the
mouse and human genomes.</p>
<p>To create an index for the GFF annotations, the <tt class="docutils literal"><span class="pre">index_gff.py</span></tt> script can be used as follows:</p>
<div class="highlight-python"><pre>python index_gff.py --index SE.mm9.gff indexed/</pre>
</div>
<p>This will index the GFF file <tt class="docutils literal"><span class="pre">SE.mm9.gff</span></tt> and output the result to the <tt class="docutils literal"><span class="pre">indexed/</span></tt> directory.
This indexed directory is then used as input to MISO.</p>
</div>
<div class="section" id="configuring-miso">
<span id="config"></span><h2><a class="toc-backref" href="#id64">Configuring MISO</a><a class="headerlink" href="#configuring-miso" title="Permalink to this headline">¶</a></h2>
<p>The default configuration file for MISO is available in <tt class="docutils literal"><span class="pre">settings/miso_settings.txt</span></tt>.
Skip to the next section if you&#8217;d like to use the default configuration file.
The default file is:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span><span class="n">data</span><span class="p">]</span>
<span class="n">filter_results</span> <span class="o">=</span> <span class="bp">True</span>
<span class="n">min_event_reads</span> <span class="o">=</span> <span class="mi">20</span>

<span class="p">[</span><span class="n">cluster</span><span class="p">]</span>
<span class="n">cluster_command</span> <span class="o">=</span> <span class="n">qsub</span>
<span class="n">long_queue_name</span> <span class="o">=</span> <span class="nb">long</span>
<span class="n">short_queue_name</span> <span class="o">=</span> <span class="n">short</span>

<span class="p">[</span><span class="n">sampler</span><span class="p">]</span>
<span class="n">burn_in</span> <span class="o">=</span> <span class="mi">500</span>
<span class="n">lag</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">num_iters</span> <span class="o">=</span> <span class="mi">5000</span>
</pre></div>
</div>
<p>The <tt class="docutils literal"><span class="pre">[data]</span></tt> section specifies whether or not events should be filtered for coverage
(<tt class="docutils literal"><span class="pre">filter_results</span></tt>) and what the minimum number of reads that a region must have for it to
be quantitated (<tt class="docutils literal"><span class="pre">min_event_reads</span></tt>). The minimum number of reads is computed over the longest genomic
region of the gene or alternatively spliced event.</p>
<p>The <tt class="docutils literal"><span class="pre">[cluster]</span></tt> section specifies parameters related to running MISO on a cluster of computers.
These are:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">cluster_command</span></tt>: the name of the program used to submit jobs to the cluster (e.g. <tt class="docutils literal"><span class="pre">qsub</span></tt>, <tt class="docutils literal"><span class="pre">bsub</span></tt>, etc.)</li>
<li><tt class="docutils literal"><span class="pre">long_queue_name</span></tt> (optional): the name of the &#8220;long queue&#8221; (jobs that take longer to finish) of the cluster queueing system.</li>
<li><tt class="docutils literal"><span class="pre">short_queue_name</span></tt> (optional): the name of the &#8220;short queue&#8221; (jobs that generally finish quicker and might get priority) of the cluster queueing system. The name of this queue is often <tt class="docutils literal"><span class="pre">short</span></tt> or <tt class="docutils literal"><span class="pre">quick</span></tt> depending on the way the cluster is configured.</li>
</ul>
<p>Both <tt class="docutils literal"><span class="pre">long_queue_name</span></tt> and <tt class="docutils literal"><span class="pre">short_queue_name</span></tt> can be omitted from the configuration file, in which
case jobs will be submitted to the cluster without specifying any queue. All parameters set in the
<tt class="docutils literal"><span class="pre">[clusters]</span></tt> section are only relevant when MISO is used on a cluster.</p>
<p>The <tt class="docutils literal"><span class="pre">[sampler]</span></tt> section specifies internal parameters related to MISO&#8217;s sampling algorithm. Most users
will not want to tweak these. The parameters are as follows:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">burn_in</span></tt>: the number of initial sampling iterations to be discarded before computing estimates.</li>
<li><tt class="docutils literal"><span class="pre">lag</span></tt>: the number of sampling iterations to skip when computing estimates. A lag of 10 would mean every 10th iteration is used.</li>
<li><tt class="docutils literal"><span class="pre">num_iters</span></tt>: the total number of sampling iterations to be computed per gene/event.</li>
</ul>
<p>The default settings of these sampling parameters was deliberately chosen to be conservative.</p>
</div>
<div class="section" id="computing-expression-estimates-psi-psi-values">
<h2><a class="toc-backref" href="#id65">Computing expression estimates (&#8220;Psi&#8221; / Ψ values)</a><a class="headerlink" href="#computing-expression-estimates-psi-psi-values" title="Permalink to this headline">¶</a></h2>
<p>To run MISO to estimate the expression level of a set of annotated isoforms or events in GFF format,
the <tt class="docutils literal"><span class="pre">run_events_analysis.py</span></tt> script is used with the <tt class="docutils literal"><span class="pre">--compute-genes-psi</span></tt> option.
This produces distributions over &#8220;Psi&#8221; values (Ψ), which stand for
&#8220;Percent Spliced In&#8221;. See Katz et. al. (2010) for a description of
these values and how they are computed.</p>
<div class="section" id="using-single-end-reads">
<h3><a class="toc-backref" href="#id66">Using single-end reads</a><a class="headerlink" href="#using-single-end-reads" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--compute-genes-psi</span></tt> takes the following arguments:</p>
<div class="highlight-python"><pre>--compute-genes-psi
                      Compute Psi values for given GFF annotation.
                      Expects two arguments: an indexed GFF
                      directory with genes to process, and a sorted, indexed
                      BAM file (with headers) to run on.</pre>
</div>
<p>For example, to run on a single-end sample:</p>
<div class="highlight-python"><pre>python run_events_analysis.py --compute-genes-psi mm9/pickled/SE data/my_sample.sorted.bam --output-dir my_sample_output/ --read-len 35</pre>
</div>
<p>This will compute expression values for the isoforms described in the indexed GFF directory <tt class="docutils literal"><span class="pre">mm9/pickled/SE</span></tt>, based on the reads in <tt class="docutils literal"><span class="pre">data/my_sample.sorted.bam</span></tt> (a sorted, indexed BAM file) an output the results to the <tt class="docutils literal"><span class="pre">my_sample_output/</span></tt> directory. Note that the <tt class="docutils literal"><span class="pre">mm9/pickled/SE</span></tt> directory in this case must be generated by the user, by running <tt class="docutils literal"><span class="pre">index_gff.py</span></tt>.</p>
<p>The <tt class="docutils literal"><span class="pre">--read-len</span></tt> option is necessary and specifies the length of the reads in the data (in this case, 35 nt.)  The directory <tt class="docutils literal"><span class="pre">mm9/pickled/SE</span></tt>, which is available in the mouse splicing event annotations, specifies an indexed version of all the skipped exon (SE) events.</p>
<p>Note that that the headers for the indexed/sorted BAM file (the files with the <tt class="docutils literal"><span class="pre">.bam.bai</span></tt> extension)
must be in the same directory as the read BAM file. In the above example, it means they must be in the
<tt class="docutils literal"><span class="pre">data/</span></tt> directory.</p>
<p>If reads were aligned to a set of splice junctions with an <em>overhang</em> constraint &#8212; i.e., a requirement
that N or more bases from each side of the junction must be covered by the read &#8212; then this can be
specified using the <tt class="docutils literal"><span class="pre">--overhang-len</span></tt> option:</p>
<div class="highlight-python"><pre>--overhang-len
                      Length of overhang constraints imposed on junctions.</pre>
</div>
</div>
<div class="section" id="using-paired-end-reads">
<h3><a class="toc-backref" href="#id67">Using paired-end reads</a><a class="headerlink" href="#using-paired-end-reads" title="Permalink to this headline">¶</a></h3>
<p>To compute expression levels using paired-end reads, use the <tt class="docutils literal"><span class="pre">--paired-end</span></tt> option:</p>
<div class="highlight-python"><pre>--paired-end
                      Run in paired-end mode. Takes mean and standard
                      deviation of insert length distribution.</pre>
</div>
<p>The insert length distribution gives the range of sizes of fragments sequenced in the
paired-end RNA-Seq run. This is used to assign reads to isoforms probabilistically.
The insert length distribution can be computed by aligning read pairs to long, constitutive
exons (like 3&#8217; UTRs) and measuring the distance between the read mates. The mean and standard
deviation of this distribution would then be given as arguments to <tt class="docutils literal"><span class="pre">--paired-end</span></tt>.
For example, to run on a paired-end sample where the mean insert length is 250 and the standard
deviation is 15, we would use: <tt class="docutils literal"><span class="pre">--paired-end</span> <span class="pre">250</span> <span class="pre">15</span></tt> when calling <tt class="docutils literal"><span class="pre">run_events_analysis.py</span></tt>
(in addition to the <tt class="docutils literal"><span class="pre">--compute-genes-psi</span></tt> option).</p>
<p>The <tt class="docutils literal"><span class="pre">--overhang-len</span></tt> option is not supported for paired-end reads.</p>
<div class="section" id="computing-the-insert-length-distribution-and-its-statistics">
<h4><a class="toc-backref" href="#id68">Computing the insert length distribution and its statistics</a><a class="headerlink" href="#computing-the-insert-length-distribution-and-its-statistics" title="Permalink to this headline">¶</a></h4>
<p>We provide a set of utilities for computing and plotting the insert length distribution in paired-end RNA-Seq samples. The insert length distribution of a sample is computed by aligning the read pairs to long constitutive exons and then measuring the insert length of each pair. The set of insert lengths obtained this way form a distribution, and summary statistics of this distribution &#8211; like its mean and standard deviation &#8211; are used by MISO to assign read pairs to isoforms.</p>
<p>The utilities in <tt class="docutils literal"><span class="pre">exon_utils.py</span></tt> and <tt class="docutils literal"><span class="pre">pe_utils.py</span></tt> can be used to first get a set of long constitituve exons to map read pairs to, and second compute the insert length distribution and its statistics.</p>
<div class="section" id="getting-a-set-of-long-constitutive-exons">
<h5><a class="toc-backref" href="#id69">Getting a set of long constitutive exons</a><a class="headerlink" href="#getting-a-set-of-long-constitutive-exons" title="Permalink to this headline">¶</a></h5>
<p>When computing the insert length distribution, it&#8217;s important to use exons that are significantly larger than the insert length that was selected for in preparing the RNA-Seq library. If the insert length selected for in preparing the RNA-Seq library was roughly 250-300 nt, we can use constitutive exons that are at least 1000 bases long, for example, so that both read pairs will map within the exon and not outside of it. The requirement that the exons be constitutive is to avoid errors in the insert length measurements that are caused by alternative splicing of the exon (which could alter the insert length.)</p>
<p>We first get all constitutive exons from a gene models GFF, like the Ensembl annotation of mouse genes (available here, <a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/gene-models/Mus_musculus.NCBIM37.65.gff.zip">Mus_musculus.NCBIM37.65.gff</a>), we use <tt class="docutils literal"><span class="pre">exon_utils.py</span></tt>:</p>
<div class="highlight-python"><pre>python exon_utils.py --get-const-exons Mus_musculus.NCBIM37.65.gff --min-exon-size 1000 --output-dir exons/</pre>
</div>
<p>This will output a GFF file (named <tt class="docutils literal"><span class="pre">Mus_musculus.NCBIM37.65.min_1000.const_exons.gff</span></tt>) into the <tt class="docutils literal"><span class="pre">exons</span></tt> directory containing only constitutive exons that are at least 1000 bases long. This file can be used to compute the insert length distribution of all mouse RNA-Seq datasets. Exons here are defined as constitutive only if they occur in all annotated transcripts of a gene.</p>
</div>
<div class="section" id="computing-the-insert-length-distribution-given-a-constitutive-exons-file">
<h5><a class="toc-backref" href="#id70">Computing the insert length distribution given a constitutive exons file</a><a class="headerlink" href="#computing-the-insert-length-distribution-given-a-constitutive-exons-file" title="Permalink to this headline">¶</a></h5>
<p>Now that we have a file containing the long constitutive exons, we can compute the insert length for any BAM file using the <tt class="docutils literal"><span class="pre">pe_utils.py</span></tt> utilities. The option <tt class="docutils literal"><span class="pre">--compute-insert-len</span></tt> takes a BAM file with the RNA-Seq sample and a GFF file containing the long constitutive exons. The insert length is defined simply as the distance between the start coordinate of the first mate in the read pair and the end coordinate of the second mate in the read pair.</p>
<p>To compute the insert length for <tt class="docutils literal"><span class="pre">sample.bam</span></tt> using the constitutive exons file we made above, run:</p>
<div class="highlight-python"><pre>python pe_utils.py --compute-insert-len sample.bam Mus_musculus.NCBIM37.65.min_1000.const_exons.gff --output-dir insert-dist/</pre>
</div>
<p>This will efficiently map the BAM reads to the exons in the GFF file using <a class="reference external" href="http://code.google.com/p/bedtools/">bedtools</a>, and use the result to compute the insert lengths. The command requires the latest version of bedtools to be installed, where the <tt class="docutils literal"><span class="pre">tagBam</span></tt> utility was added the ability to output the interval that each read maps to in its BAM output file. The reads that aligned to these exons will be kept in a directory called <tt class="docutils literal"><span class="pre">bam2gff_Mus_musculus.NCBIM37.65.min_1000.const_exons.gff/</span></tt> in the <tt class="docutils literal"><span class="pre">insert-dist</span></tt> directory.</p>
<p>This will generate an insert length distribution file (ending in <tt class="docutils literal"><span class="pre">.insert_len</span></tt>) that is tab-delimited with two columns, the first specifying the exon that the read pairs were aligned to, and the second specifying a comma-separated list of insert lengths obtained from aligning the read pairs to that exon. The number of entries in the file corresponds to the number of exons from the original GFF that were used in computing the insert length distribution.</p>
<p>The first line (header) of the the insert length file gives the mean, standard deviation and dispersion values for the distribution, for example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c">#mean=129.0,sdev=12.1,dispersion=1.1,num_pairs=862148</span>
</pre></div>
</div>
<p>This indicates that the mean of the insert length distribution was 129, the standard deviation was 12.1, dispersion was 1.1, and the number of read pairs used to estimate the distribution and compute these values was 862,148.</p>
<p>The dispersion constant <em>d</em> is defined as the standard deviation σ of the insert length distribution divided by the square root of its mean μ:</p>
<blockquote>
<div><div class="align-center figure align-center">
<a class="reference internal image-reference" href="_images/dispersion.png"><img alt="Definition of dispersion constant for an insert length distribution" src="_images/dispersion.png" style="width: 92.8px; height: 38.4px;" /></a>
<p class="caption"><em>Dispersion constant d of an insert length distribution</em></p>
</div>
</div></blockquote>
<p>Intuitively, the dispersion constant <em>d</em> corresponds to how variable the insert length distribution is about its mean. The lower the <em>d</em> value is, the tighter/more precise the insert length distribution is in the RNA-Seq sample. The dispersion constant and an interpretation of its values is discussed in the MISO paper (see <a class="reference external" href="http://www.nature.com/nmeth/journal/v7/n12/full/nmeth.1528.html">Analysis and design of RNA sequencing experiments for identifying isoform regulation</a>).</p>
<p>The insert length distribution can be plotted using the <tt class="docutils literal"><span class="pre">.insert_len</span></tt> file using <a class="reference external" href="sashimi.html">sashimi_plot</a>.</p>
<p>To compute the insert length by simple pairing of read mates by their ID (assuming that mates belonging to a pair have the same read ID in the BAM), explicitly ignoring the relevant BAM fields specifying whether or not the reads are properly paired, use the <tt class="docutils literal"><span class="pre">--no-bam-filter</span></tt> option.</p>
<blockquote>
<div><div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The option <tt class="docutils literal"><span class="pre">--compute-insert-len</span></tt> can take a comma-separated list of BAM filenames in case you want to compute the insert length for multiple samples in one command. The <tt class="docutils literal"><span class="pre">--compute-insert-len</span></tt> option of <tt class="docutils literal"><span class="pre">pe_utils</span></tt> by default uses only exons from the GFF file that are 500 bases or longer. This can be tweaked by passing <tt class="docutils literal"><span class="pre">pe_utils</span></tt> the optional <tt class="docutils literal"><span class="pre">--min-exon-size</span> <span class="pre">N</span></tt> argument, which will only use exons of size <tt class="docutils literal"><span class="pre">N</span></tt> or longer. In our example, we used a constitutive exons GFF file that contained only exons that are 100 bp or longer, so the default settings will consider all exons in that file.</p>
</div>
</div></blockquote>
</div>
</div>
</div>
<div class="section" id="distributing-miso-jobs-to-a-cluster">
<h3><a class="toc-backref" href="#id71">Distributing MISO jobs to a cluster</a><a class="headerlink" href="#distributing-miso-jobs-to-a-cluster" title="Permalink to this headline">¶</a></h3>
<p>Running MISO on a cluster is <strong>highly recommended</strong>. Since each gene or alternative splicing event
can be treated independently, the computation of expression levels is easily parallelized on a cluster.
This can reduce run times significantly, from days to hours on typical data sets. To run MISO on the
cluster, the following options can be used:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">--use-cluster</span></tt>: causes jobs to be sent to a cluster</li>
<li><tt class="docutils literal"><span class="pre">--chunk-jobs</span> <span class="pre">N</span></tt> (optional): places <tt class="docutils literal"><span class="pre">N</span></tt> many genes or events into each job. For example, if running MISO on a set of skipped exons, <tt class="docutils literal"><span class="pre">--chunk-jobs</span> <span class="pre">200</span></tt> will place roughly 200 exons whose expression is to be estimated into each job. If each job was sent to a separate node, it would mean every node in parallel will sequentially compute the expression level of 200 exons. The value of <tt class="docutils literal"><span class="pre">--chunk-jobs</span></tt> depends on both cluster size/availability and on the number of exons whose expression is to be estimated. On very large clusters where nodes are generally available, a relatively small setting of <tt class="docutils literal"><span class="pre">--chunk-jobs</span></tt> is likely to be more efficient.</li>
</ul>
<p>MISO will use the cluster submission and queue settings described in the settings file to submit jobs (see <a class="reference internal" href="sashimi.html#config"><em>Visualizing RNA-Seq reads along isoforms with MISO estimates</em></a>). An example of a MISO pipeline for computing isoform expression estimates and detecting differentially expressed isoforms is given in <a class="reference internal" href="#pipeline"><em>Example MISO pipeline</em></a>.</p>
<p>To use MISO on Sun Grid Engine, the following options to <tt class="docutils literal"><span class="pre">run_events_analysis.py</span></tt> can be used:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">--SGEarray</span></tt>: Run jobs on Sun Grid Engine. To be used in conjunction with <tt class="docutils literal"><span class="pre">--use-cluster</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">--job-name</span></tt>: Name the job submitted to Sun Grid Engine system.</li>
</ul>
<p>Thanks to Michael Lovci for this feature.</p>
</div>
<div class="section" id="raw-miso-output-format">
<h3><a class="toc-backref" href="#id72">Raw MISO output format</a><a class="headerlink" href="#raw-miso-output-format" title="Permalink to this headline">¶</a></h3>
<p>When running MISO, through <tt class="docutils literal"><span class="pre">run_events_analysis.py</span> <span class="pre">--compute-genes-psi</span></tt>, the raw output will be a set of posterior distributions over Ψ values. Each exon or gene, depending on whether the analysis is exon or isoform-centric, will have its own file containing posterior samples from the distribution over Ψ.</p>
<p>Most users will never need to access these raw distributions and can simply summarize this raw output as described in <a class="reference internal" href="#summarizing"><em>Summarizing MISO output</em></a>, or perform a comparison between a set of samples as described in <a class="reference internal" href="#comparing-samples"><em>Detecting differentially expressed isoforms</em></a>.</p>
<p>Each gene or exon will have a corresponding distribution file (ending in <tt class="docutils literal"><span class="pre">.miso</span></tt>), split into subdirectories by chromosome. The resulting directory structure might be:</p>
<div class="highlight-python"><pre>my_sample_output/
  chr1/
  chr2/
  ...
  chrX/</pre>
</div>
<p>The distribution file for an event, e.g. <tt class="docutils literal"><span class="pre">chr1/my_exon.miso</span></tt>, will contain tab-separated columns that specify the sampled Ψ values and the log score under the MISO model for each sample, e.g.:</p>
<div class="highlight-python"><pre>#isoforms=['TandemUTRCore_TandemUTRExt','TandemUTRCore']      exon_lens=('TandemUTRCore',1986),('TandemUTRExt',1021)  iters=5000      burn_in=500     lag=10  percent_accept=92.90    proposal_type=drift
sampled_psi   log_score
0.0691,0.9309 -989.1296
0.0692,0.9308 -1002.1173
0.0595,0.9405 -994.5029
0.0587,0.9413 -988.9871
0.0539,0.9461 -980.4465
0.0571,0.9429 -986.1921
0.0482,0.9518 -980.2666
0.0392,0.9608 -999.4088</pre>
</div>
<p>The header line (beginning with <tt class="docutils literal"><span class="pre">#</span></tt>) gives information about the gene/event in the file, like the names of the isoforms in the annotation and the length of the exons, as well as values of internal MISO parameters (like burn-in or lag, described in <a class="reference internal" href="sashimi.html#config"><em>Visualizing RNA-Seq reads along isoforms with MISO estimates</em></a>).</p>
<p>The first column (<tt class="docutils literal"><span class="pre">sampled_psi</span></tt>) lists the Ψ value sampled and the second column (<tt class="docutils literal"><span class="pre">log_score</span></tt>) gives the log score of that sample under the MISO model. The Ψ value is comma-separated and will have as many entries as there are isoforms. In the above file, there are two isoforms, and so two entries in the Ψ column that sum to 1.</p>
</div>
</div>
</div>
<div class="section" id="summarizing-miso-output">
<span id="summarizing"></span><h1><a class="toc-backref" href="#id73">Summarizing MISO output</a><a class="headerlink" href="#summarizing-miso-output" title="Permalink to this headline">¶</a></h1>
<p>To summarize MISO output and obtain confidence intervals (CIs) for Ψ values, the <tt class="docutils literal"><span class="pre">run_miso.py</span> <span class="pre">--summarize-samples</span></tt> option can be used:</p>
<div class="highlight-python"><pre>--summarize-samples
                       Compute summary statistics of the given set of
                       samples. Expects a directory with MISO output and
                       a directory to output summary file to.</pre>
</div>
<p>For example, to summarize the MISO output in <tt class="docutils literal"><span class="pre">my_sample_output/</span></tt>, run:</p>
<div class="highlight-python"><pre>python run_miso.py --summarize-samples my_sample_output/ summary_output/</pre>
</div>
<p>This will create a subdirectory <tt class="docutils literal"><span class="pre">summary_output/summary/</span></tt> with a file <tt class="docutils literal"><span class="pre">my_sample_output.miso_summary</span></tt> that summarizes all the events in the sample in <tt class="docutils literal"><span class="pre">my_sample_output/</span></tt>.</p>
<div class="section" id="summary-file-output-format">
<span id="summary-format"></span><h2><a class="toc-backref" href="#id74">Summary file output format</a><a class="headerlink" href="#summary-file-output-format" title="Permalink to this headline">¶</a></h2>
<p>The summary file is a tab-separated format containing the following columns:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">event_name</span></tt>: name of the event or gene from the input annotation,</li>
<li><tt class="docutils literal"><span class="pre">miso_posterior_mean</span></tt>: an estimate of Ψ (derived by taking the mean of the posterior distribution over Ψ),</li>
<li><tt class="docutils literal"><span class="pre">ci_low</span></tt>: the low bound of the 95% confidence interval on the Ψ estimate,</li>
<li><tt class="docutils literal"><span class="pre">ci_high</span></tt>: the high bound of the 95% confidence interval on the Ψ estimate.</li>
<li><tt class="docutils literal"><span class="pre">isoforms</span></tt>: the names of the isoforms that make up the event/gene
quantitated (from the input annotations)</li>
<li><tt class="docutils literal"><span class="pre">counts</span></tt>: raw read counts that support each isoform. Lists the number of reads that support each combination of isoforms. Each read type is represented as a comma-separated list (enclosed in parentheses) with number of entries equal to the number of isoforms quantitated, and a 1 or 0 in each entry depending on whether the read was consistent or inconsistent with the isoform. For example, for an alternative event with two isoforms, the entry might be: <tt class="docutils literal"><span class="pre">(1,0):28,(0,1):50,(1,1):12</span></tt>, where <tt class="docutils literal"><span class="pre">0</span></tt> and <tt class="docutils literal"><span class="pre">1</span></tt> correspond to the first and second isoforms. This indicates that 28 reads support isoform 1 but not 2, while 50 reads support the second but not the first, and 12 reads support both isoforms. The field <tt class="docutils literal"><span class="pre">(0,0):N</span></tt>, if outputted, indicates the number of reads <tt class="docutils literal"><span class="pre">N</span></tt> that mapped to the region but were thrown out (e.g. because they are not consistent with the isoforms in the annotation, violate overhang constraints, or other related reasons.)</li>
<li><tt class="docutils literal"><span class="pre">assigned_counts</span></tt>: inferred assignment of reads to isoforms. Lists each isoform (0-based) and the number of reads assigned to it. For example, for an alternative event with two isoforms, the entry might be: <tt class="docutils literal"><span class="pre">0:330,1:34</span></tt>, indicating that 330 reads were assigned to the first isoform (isoform 0) and 34 reads were assigned to the second isoform (isoform 1).</li>
</ul>
<p>Starting with release <tt class="docutils literal"><span class="pre">misopy-0.4.1</span></tt>, three additional columns are outputted with information about the gene/event (which are read from the input GFF annotation):</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">chrom</span></tt>: chromosome name</li>
<li><tt class="docutils literal"><span class="pre">strand</span></tt>: strand of gene/event</li>
<li><tt class="docutils literal"><span class="pre">mRNA_starts</span></tt>: comma-separated list of start genomic coordinates for <tt class="docutils literal"><span class="pre">mRNA</span></tt> entries in gene/event</li>
<li><tt class="docutils literal"><span class="pre">mRNA_ends</span></tt>: comma-separated list of end genomic coordinates for <tt class="docutils literal"><span class="pre">mRNA</span></tt> enries in gene/event</li>
</ul>
<p>When one of those fields is not available (e.g. <tt class="docutils literal"><span class="pre">strand</span></tt>), <tt class="docutils literal"><span class="pre">NA</span></tt> is outputted in its place.</p>
<p>Note that <tt class="docutils literal"><span class="pre">miso_posterior_mean</span></tt> is an estimate of Ψ and [<tt class="docutils literal"><span class="pre">ci_low</span></tt>, <tt class="docutils literal"><span class="pre">ci_high</span></tt>] are the 95% confidence limits of that estimate.</p>
</div>
</div>
<div class="section" id="detecting-differentially-expressed-isoforms">
<span id="comparing-samples"></span><h1><a class="toc-backref" href="#id75">Detecting differentially expressed isoforms</a><a class="headerlink" href="#detecting-differentially-expressed-isoforms" title="Permalink to this headline">¶</a></h1>
<p>Once MISO output has been computed for two samples or more, pairwise comparison of differential
isoform expression can be computed. To test if an isoform or exon is differentially expressed between
between samples, the <em>Bayes factor</em> is computed for the exon, which represents the weight of the evidence
in the data in favor of differential expression versus not. For example, a Bayes factor of 2 would mean
that the isoform/exon is two times more likely to be differentially expressed than not.
The computation of Bayes factors is described in detail in Katz et. al. (2010).</p>
<p>To compute Bayes factors, the <tt class="docutils literal"><span class="pre">run_miso.py</span> <span class="pre">--compare-samples</span></tt> option can be used. For example:</p>
<div class="highlight-python"><pre>python run_miso.py --compare-samples control/ knockdown/ comparisons/</pre>
</div>
<p>This would compare the MISO output in <tt class="docutils literal"><span class="pre">control/</span></tt> to <tt class="docutils literal"><span class="pre">knowndown/</span></tt>, and output the result
to the directory <tt class="docutils literal"><span class="pre">comparisons/control_vs_knockdown</span></tt>. Note that
&#8220;<tt class="docutils literal"><span class="pre">vs</span></tt>&#8221; is used to denote sample comparisons. The output format of
this comparison is described below.</p>
<div class="section" id="output-of-samples-comparison">
<h2><a class="toc-backref" href="#id76">Output of samples comparison</a><a class="headerlink" href="#output-of-samples-comparison" title="Permalink to this headline">¶</a></h2>
<p>The output of <tt class="docutils literal"><span class="pre">--compare-samples</span></tt> will be in two subdirectories:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">bayes-factors/</span></tt>: contains the file summarizing the output with the Bayes factors (ending in <tt class="docutils literal"><span class="pre">.miso_bf</span></tt>)</li>
<li><tt class="docutils literal"><span class="pre">delta-posteriors/</span></tt>: contains the posterior distributiosn over the change in Psi (Δ Ψ  values) for each event</li>
</ul>
<p>The main output is in <tt class="docutils literal"><span class="pre">bayes-factors/</span></tt> and most users will never need to directly
look at the contents of <tt class="docutils literal"><span class="pre">delta-posteriors/</span></tt>, though its format is described in <a class="reference internal" href="#delta-posteriors"><em>Raw Δ Ψ distribution output format</em></a>.</p>
<div class="section" id="bayes-factors-file-output-format">
<span id="bf-output"></span><h3><a class="toc-backref" href="#id77">Bayes factors file output format</a><a class="headerlink" href="#bayes-factors-file-output-format" title="Permalink to this headline">¶</a></h3>
<p>The Bayes factors file (ending in <tt class="docutils literal"><span class="pre">.miso_bf</span></tt>) will be stored in the <tt class="docutils literal"><span class="pre">bayes-factors/</span></tt> directory generated
by the <tt class="docutils literal"><span class="pre">--compare-samples</span></tt> option. The Bayes factor file is a tab-separated format containing the following columns:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">event_name</span></tt>: name of the event from the input annotation file,</li>
<li><tt class="docutils literal"><span class="pre">sample1_posterior_mean</span></tt>: Estimate of Ψ (mean of posterior distribution) for sample 1,</li>
<li><tt class="docutils literal"><span class="pre">sample1_ci_low</span></tt>: Low bound of 95% confidence interval for sample 1 Ψ estimate,</li>
<li><tt class="docutils literal"><span class="pre">sample1_ci_high</span></tt>: High bound of the 95% confidence interval for sample 1 Ψ estimate,</li>
<li><tt class="docutils literal"><span class="pre">sample2_posterior_mean</span></tt>: Estimate of Ψ (mean of posterior distribution) for sample 2,</li>
<li><tt class="docutils literal"><span class="pre">sample2_ci_low</span></tt>: Low bound of 95% confidence interval for sample 2 Ψ estimate,</li>
<li><tt class="docutils literal"><span class="pre">sample2_ci_high</span></tt>: High bound of 95% confidence interval for sample 2 Ψ estimate,</li>
<li><tt class="docutils literal"><span class="pre">diff</span></tt>: The difference (Δ Ψ) between sample 1 and sample 2 Ψ estimates, i.e. <tt class="docutils literal"><span class="pre">sample1_posterior_mean</span> <span class="pre">-</span> <span class="pre">sample2_posterior_mean</span></tt></li>
<li><tt class="docutils literal"><span class="pre">bayes_factor</span></tt>: Bayes factor for the comparison, corresponding to the odds of differential expression ,where the change in Ψ is non-zero) over no differential expression (where the change is zero),</li>
<li><tt class="docutils literal"><span class="pre">isoforms</span></tt>: the names of the transcripts or events that were quantitated (based on the input annotation),</li>
<li><tt class="docutils literal"><span class="pre">sample1_counts</span></tt>: raw counts for sample 1 for each read type (same format as the <tt class="docutils literal"><span class="pre">counts</span></tt> field described in <a class="reference internal" href="#summary-format"><em>Summary file output format</em></a>),</li>
<li><tt class="docutils literal"><span class="pre">sample1_assigned_counts</span></tt>: inferred assignment of reads to isoforms in sample 1 (same format as the <tt class="docutils literal"><span class="pre">assigned_counts</span></tt> field described in <a class="reference internal" href="#summary-format"><em>Summary file output format</em></a>),</li>
<li><tt class="docutils literal"><span class="pre">sample2_counts</span></tt>: raw counts for sample 2 for each read type,</li>
<li><tt class="docutils literal"><span class="pre">sample2_assigned_counts</span></tt>: inferred assignment of reads to isoforms in sample 2</li>
</ul>
<p>If MISO is run on an annotation that specifies more than two isoforms per event/gene, then the fields for the Ψ estimates become comma-separated and give the estimates for each isoform. For example, in the case of three isoforms, <tt class="docutils literal"><span class="pre">sample1_ci_low</span></tt> would be a comma-separated list of the lower bounds of the first, second and third isoform, and <tt class="docutils literal"><span class="pre">sample1_ci_high</span></tt> would be a comma-separated list of the upper bounds of the first, second and third isoform, and similarly for the fields involving the other sample.</p>
<p>Starting with release <tt class="docutils literal"><span class="pre">misopy-0.4.1</span></tt>, three additional columns are outputted with information about the gene/event (which are read from the input GFF annotation):</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">chrom</span></tt>: chromosome name</li>
<li><tt class="docutils literal"><span class="pre">strand</span></tt>: strand of gene/event</li>
<li><tt class="docutils literal"><span class="pre">mRNA_starts</span></tt>: comma-separated list of start genomic coordinates for <tt class="docutils literal"><span class="pre">mRNA</span></tt> entries in gene/event</li>
<li><tt class="docutils literal"><span class="pre">mRNA_ends</span></tt>: comma-separated list of end genomic coordinates for <tt class="docutils literal"><span class="pre">mRNA</span></tt> enries in gene/event</li>
</ul>
<p>When one of those fields is not available (e.g. <tt class="docutils literal"><span class="pre">strand</span></tt>), <tt class="docutils literal"><span class="pre">NA</span></tt> is outputted in its place.</p>
</div>
<div class="section" id="raw-delta-psi-distribution-output-format">
<span id="delta-posteriors"></span><h3><a class="toc-backref" href="#id78">Raw Δ Ψ distribution output format</a><a class="headerlink" href="#raw-delta-psi-distribution-output-format" title="Permalink to this headline">¶</a></h3>
<p>The output of the <tt class="docutils literal"><span class="pre">--compare-samples</span></tt> option generates not only Bayes factors but also posterior
distributions over the Δ Ψ value of each isoform. These files are in the <tt class="docutils literal"><span class="pre">delta-posteriors/</span></tt>
subdirectory of the specified output directory. Each alternative event will have its own Δ Ψ file, split into batches. The resulting directory structure (using the above example of a call to <tt class="docutils literal"><span class="pre">--compare-samples</span></tt>) would be:</p>
<div class="highlight-python"><pre>comparisons/
  control_vs_knockdown/
    batch_500_1/
    batch_500_2/
    ...</pre>
</div>
<p>where <tt class="docutils literal"><span class="pre">batch_500_1</span></tt> contains the Δ Ψ files for the first set of 500 events in the sample, <tt class="docutils literal"><span class="pre">batch_500_2</span></tt> for the second set of 500 events in the sample, and so on. Within each batch directory, there will be a file per event (ending in <tt class="docutils literal"><span class="pre">.miso_dp</span></tt>) that contains the Δ Ψ distribution for that event. Each file simply contains a single column labeled <tt class="docutils literal"><span class="pre">delta_posteriors</span></tt> where each row is a sampled Δ Ψ value. The values in each file form an estimate of the posterior distribution over Δ Ψ for the event.</p>
</div>
</div>
</div>
<div class="section" id="example-miso-pipeline">
<span id="pipeline"></span><h1><a class="toc-backref" href="#id79">Example MISO pipeline</a><a class="headerlink" href="#example-miso-pipeline" title="Permalink to this headline">¶</a></h1>
<p>Below is an example of a MISO pipeline, where Ψ values for a set of alternative events are computed (with confidence intervals) for a pair of samples called &#8220;control&#8221; and &#8220;knockdown&#8221;. A samples comparison is performed to detect differentially expressed isoforms between the samples.</p>
<div class="highlight-python"><pre>## Run MISO on a pair of paired-end sample (with insert length distribution with mean 250,
## standard deviation 15) using the mouse genome skipped exon annotations using the
## the cluster

# Compute Psi values for control sample
python run_events_analysis.py --compute-genes-psi mm9/pickled/SE data/control.bam --output-dir SE/control/ --read-len 35 --paired-end 250 15 --use-cluster

# Compute Psi values for knockdown sample
python run_events_analysis.py --compute-genes-psi mm9/pickled/SE data/knockdown.bam --output-dir SE/knockdown/ --read-len 35 --paired-end 250 15 --use-cluster


## Summarize the output (only run this once --compute-genes-psi finished!)
## This will create a "summary" directory in SE/control/ and in SE/knockdown/
python run_miso.py --summarize-samples SE/control/ SE/control/
python run_miso.py --summarize-samples SE/knockdown/ SE/knockdown/

## Detect differentially expressed isoforms between "control" and "knockdown"
## This will compute Bayes factors and delta Psi values between the samples
## and place the results in the directory SE/comparisons/control_vs_knockdown
python run_miso.py --compare-samples SE/control/ SE/knockdown/ SE/comparisons/</pre>
</div>
</div>
<div class="section" id="interpreting-and-filtering-miso-output">
<span id="interpreting"></span><h1><a class="toc-backref" href="#id80">Interpreting and filtering MISO output</a><a class="headerlink" href="#interpreting-and-filtering-miso-output" title="Permalink to this headline">¶</a></h1>
<p>Once MISO runs are completed and pairwise comparisons between your samples have been made to compute Bayes factors (using <tt class="docutils literal"><span class="pre">--compare-samples</span></tt>), the resulting files can be interpreted to detect differential events.  For each pairwise comparison, we will have a <tt class="docutils literal"><span class="pre">.miso_bf</span></tt> file which contains the Ψ values of each event in both samples along with confidence intervals (assuming the event is detectable in both), the Δ Ψ values, the Bayes factor, and other useful information about the read counts used to compute these values.</p>
<p>MISO comes with several utilities (such as <tt class="docutils literal"><span class="pre">filter_events.py</span></tt>, described in <a class="reference internal" href="#filterscript"><em>Filtering differentially expressed events</em></a>) for helping with the interpretation and analysis of these output files. This section serves as a brief guide the MISO output files, the utilities for analyzing them, and to various caveats in interpreting the output.</p>
<div class="section" id="using-the-read-class-counts">
<span id="usingcounts"></span><h2><a class="toc-backref" href="#id81">Using the read class counts</a><a class="headerlink" href="#using-the-read-class-counts" title="Permalink to this headline">¶</a></h2>
<p>A first-pass filter for detecting differentially changing events can apply a Bayes factor cutoff along with a Δ Ψ cutoff, using the <tt class="docutils literal"><span class="pre">bayes_factor</span></tt> and <tt class="docutils literal"><span class="pre">diff</span></tt> fields of the samples comparison output file. Section <a class="reference internal" href="#filterscript"><em>Filtering differentially expressed events</em></a> describes a utility that will apply these filters for exon-centric analyses.</p>
<p>It is critical to note that the default MISO settings were intentionally set to be highly inclusive, and therefore do not impose any stringent coverage criteria on the events for which Ψ values and Bayes factors are computed.</p>
<blockquote>
<div><div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">By default, MISO requires only 20 reads to be present in the locus/event of interest. For many events, only a tiny fraction of these reads might be in informative regions &#8211; like junctions or regions unique to one of the isoforms &#8211; and so the expression levels of these events might not be reliably estimated. It is important to use the information provided by MISO about the raw number of read counts used in estimating the expression levels of each event/gene.</p>
</div>
</div></blockquote>
<p>In addition to Ψ values, confidence intervals and Bayes factors, the output files of MISO (such as summary files or sample comparison files) also contain the number of read counts informative about each isoform.
The relevant read counts are given in the <tt class="docutils literal"><span class="pre">counts</span></tt> field of the <tt class="docutils literal"><span class="pre">.miso,</span> <span class="pre">.summary,</span> <span class="pre">.miso_bf</span></tt> files &#8211; its format is described in detail in <a class="reference internal" href="#summary-format"><em>Summary file output format</em></a>). For the two-isoform case in exon-centric analyses, the counts field general format is:</p>
<div class="highlight-python"><pre>(1,0):X,(0,1):Y,(1,1):Z,(0,0):L</pre>
</div>
<p>where <tt class="docutils literal"><span class="pre">X,</span> <span class="pre">Y,</span> <span class="pre">Z,</span> <span class="pre">L</span></tt> are integer counts corresponding to the number of reads in each of these categories. Class <tt class="docutils literal"><span class="pre">(1,0)</span></tt> are reads consistent with the first isoform in the annotation but not the second, class <tt class="docutils literal"><span class="pre">(0,1)</span></tt> are reads consistent with the second but not the first, class <tt class="docutils literal"><span class="pre">(1,1)</span></tt> are consistent with both isoforms, and reads in <tt class="docutils literal"><span class="pre">(0,0)</span></tt> are consistent with neither. These are illustrated graphically in the figure below for alternatively skipped exon event:</p>
<blockquote>
<div><div class="align-center figure align-center">
<a class="reference internal image-reference" href="_images/read_classes_for_SE.png"><img alt="Illustration of MISO read class counts for a skipped exon" src="_images/read_classes_for_SE.png" style="width: 320.0px; height: 292.0px;" /></a>
<p class="caption"><em>Illustration of MISO output read class counts for a skipped exon</em></p>
</div>
</div></blockquote>
<p>The class <tt class="docutils literal"><span class="pre">(0,0)</span></tt> denotes reads that were incompatible with either isoform, either because they didn&#8217;t match these isoforms or were otherwise unusable (e.g. they were part of a read-pair where one end was unmappable, in the case of paired-end reads.) Reads in this class are not used by MISO &#8211; they are only reported to indicate that they were discarded.</p>
<blockquote>
<div><div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Events that meet the very minimal default coverage criteria do not necessarily have sufficient read support to be considered alternatively spliced in your RNA-Seq samples.</p>
</div>
</div></blockquote>
<p>In practice, events that might not be alternatively spliced in the samples of interest meet the default minimal read coverage criteria, although their read class counts imply that they are not reliable estimates. Such a case might be an alternative exon in the input annotation to MISO that has, for example, 19 reads in the exon upstream of the skipped exon, and 1 read in the exon downstream of the skipped exon. This event would have a <tt class="docutils literal"><span class="pre">counts</span></tt> signature of: <tt class="docutils literal"><span class="pre">(1,1):20,(1,0):0,(0,1):0</span></tt>. The <tt class="docutils literal"><span class="pre">(1,0)</span></tt> and <tt class="docutils literal"><span class="pre">(0,1)</span></tt> reads, which are most informative about the relative inclusion levels of the exon, are both 0, so unless these are non-zero in other samples in the dataset, the estimate of the inclusion levels of this exon will not be reliable.</p>
<p>For this reason, <strong>coverage filters</strong> must be imposed by the user when post-processing MISO output. An example of such a filter is given below for exon-centric analyses.</p>
<blockquote>
<div><div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p>A commonly-used <strong>coverage filter</strong> in exon-centric (two-isoform) analyses requires the following minimal counts criteria to be met:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">X</span> <span class="o">+</span> <span class="n">Y</span> <span class="o">&gt;=</span> <span class="n">N</span><span class="p">,</span> <span class="n">Y</span> <span class="o">&gt;=</span> <span class="mi">1</span>
</pre></div>
</div>
<p class="last">in <em>at least one</em> of the RNA-Seq samples processed by MISO, where <tt class="docutils literal"><span class="pre">X,</span> <span class="pre">Y</span></tt> are the number of reads in the <tt class="docutils literal"><span class="pre">(1,0)</span></tt> and <tt class="docutils literal"><span class="pre">(0,1)</span></tt> read classes, respectively, and <tt class="docutils literal"><span class="pre">N</span></tt> is an arbitrary but sizeable number (e.g. 10 or 20). This filter requires that the sum of inclusion and exclusion-supporting reads be greater than or equal to <tt class="docutils literal"><span class="pre">N</span></tt>, and that the read class supporting exclusion is non-zero in at least one of the samples. This guarantees that at least some of the reads are informative about the inclusive or exclusive isoform. Note that this filter does not guarantee that there will be junction evidence for the inclusion isoform, but does guarantee that there will be a junction evidence for skipping of the exon.</p>
</div>
</div></blockquote>
<p>To increase power to detect differential events, coverage filters should be applied in aggregate across all RNA-Seq samples that were processed by MISO. To detect <em>switch-like</em> exons that are fully included in one sample and fully excluded in another, we should require the number of exon exclusion supporting reads be non-zero in at least one of the samples &#8211; or summed across samples &#8211; rather than in every sample. This filter can be easily applied by parsing the <tt class="docutils literal"><span class="pre">counts</span></tt> field in the samples comparison (<tt class="docutils literal"><span class="pre">.miso_bf</span></tt>) file produced by MISO for each pairwise comparison.</p>
<p>For isoform-centric analyses where the number of read classes is considerably larger, MISO will output the counts in the same format. For a gene with 3 isoforms, we can have the read classes <tt class="docutils literal"><span class="pre">(1,0,0),(0,1,0),(0,0,1),...,(1,1,1)</span></tt> and other filters can be applied, depending on the goal of the analysis, to select genes that have sufficient coverage to be reliably estimated.</p>
</div>
<div class="section" id="using-the-confidence-intervals">
<span id="confidence-intervals"></span><h2><a class="toc-backref" href="#id82">Using the confidence intervals</a><a class="headerlink" href="#using-the-confidence-intervals" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><div class="align-right figure align-right">
<a class="reference internal image-reference" href="_images/confidence_intervals.png"><img alt="MISO confidence intervals" src="_images/confidence_intervals.png" style="width: 165.2px; height: 213.5px;" /></a>
</div>
</div></blockquote>
<p>Important information is contained in the confidence intervals outputted by MISO for each estimate of Ψ. The width of the interval corresponds to our confidence in the estimate of the Ψ value, and so the larger the confidence width is, the less reliable the estimate. As expected, the width of the confidence interval will depend on the number of informative reads, such as reads crossing junctions or reads in regions that uniquely identify the isoform.</p>
<p>The figure on the right shows two posterior distributions over Ψ for the same alternative splicing event in two samples, where the exon had low coverage (i.e. small number of reads) in both. The posterior distribution is plotted as a black histogram, whose mean is indicated by the red vertical line and 95% confidence intervals shown by dotted grey vertical lines. The actual Ψ values and the confidence interval bounds are given to the right of the distribution. In the top distribution, the confidence interval width is <tt class="docutils literal"><span class="pre">.64</span></tt> (since <tt class="docutils literal"><span class="pre">.88</span> <span class="pre">-</span> <span class="pre">.24</span> <span class="pre">=</span> <span class="pre">.64</span></tt>) which is extremely wide. The distribution&#8217;s probable range spans from 20 to nearly 90%, far too broad to be useful. In the bottom distribution, the confidence intervals are narrower, though still considerable. Because of the large amount of uncertainty in the top distribution, we cannot confidently conclude that the true Ψ value is likely to be different between the two. Although the difference of means between the distributions is large (<tt class="docutils literal"><span class="pre">.53</span> <span class="pre">-</span> <span class="pre">.2</span> <span class="pre">=</span> <span class="pre">.33</span></tt>), the confidence intervals overlap considerably as a result of the uncertainty in the top distribution. In practice, requiring no overlap at all between the confidence intervals of two samples is too restrictive, but the width of each estimate&#8217;s confidence interval and the relationship between the intervals should be kept in mind when interpreting the results.</p>
</div>
<div class="section" id="filtering-differentially-expressed-events">
<span id="filterscript"></span><h2><a class="toc-backref" href="#id83">Filtering differentially expressed events</a><a class="headerlink" href="#filtering-differentially-expressed-events" title="Permalink to this headline">¶</a></h2>
<p>Given a MISO Bayes factor comparison file for two-isoform events, events can be filtered based on their coverage or magnitude of change. The <tt class="docutils literal"><span class="pre">filter_events.py</span></tt> script allows filtering of events, based on the following criteria:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">--num-total</span> <span class="pre">N</span></tt>: Number of total reads aligning to any isoform (or to both isoforms) has to be greater than or equal to N.</li>
<li><tt class="docutils literal"><span class="pre">--num-inc</span> <span class="pre">N</span></tt>: Number of inclusion reads (i.e. reads supporting the first isoform) has to be greater than or equal to N.</li>
<li><tt class="docutils literal"><span class="pre">--num-exc</span> <span class="pre">N</span></tt>: Number of exclusion reads (i.e. reads supporting the second isoform) has to be greater than or equal to N.</li>
<li><tt class="docutils literal"><span class="pre">--num-sum-inc-exc</span> <span class="pre">N</span></tt>: The sum of inclusion and exclusion reads has to be greater than or equal to N.</li>
<li><tt class="docutils literal"><span class="pre">--delta-psi</span> <span class="pre">P</span></tt>: The absolute Δ Ψ value must be greater than or equal to P (where P is in [0, 1]).</li>
<li><tt class="docutils literal"><span class="pre">--bayes-factor</span> <span class="pre">N</span></tt>: The Bayes factor must be greater than or equal to N.</li>
<li><tt class="docutils literal"><span class="pre">--apply-both</span></tt>: By default, the above filters on read coverage of the isoforms are required to be true in only one of the samples. To consider only events where the filters apply in both samples, use this option.</li>
</ul>
<p>For example, to filter the file <tt class="docutils literal"><span class="pre">control.miso_bf</span></tt> to contain only events with: (a) at least 1 inclusion read, (b) 1 exclusion read, such that (c) the sum of inclusion and exclusion reads is at least 10, and (d) the Δ Ψ is at least 0.20 and (e) the Bayes factor is at least 10, and (a)-(e) are true in one of the samples, use the following command:</p>
<div class="highlight-python"><pre>python filter_events.py --filter control.miso_bf --num-inc 1 --num-exc 1 --num-sum-inc-exc 10 --delta-psi 0.20 --bayes-factor 10 --output-dir filtered/</pre>
</div>
<p>This will output a filtered Bayes factor file, <tt class="docutils literal"><span class="pre">filtered/control.miso_bf.filtered</span></tt> that contains only events meeting the above criteria.</p>
</div>
</div>
<div class="section" id="visualizing-and-plotting-miso-output">
<span id="plotting"></span><h1><a class="toc-backref" href="#id84">Visualizing and plotting MISO output</a><a class="headerlink" href="#visualizing-and-plotting-miso-output" title="Permalink to this headline">¶</a></h1>
<p>MISO comes with a built-in utility, <a class="reference external" href="sashimi.html">sashimi_plot</a>, for visualizing its output and for plotting raw RNA-Seq read densities along exons and junctions.</p>
</div>
<div class="section" id="frequently-asked-questions-faq">
<span id="faq"></span><h1><a class="toc-backref" href="#id85">Frequently Asked Questions (FAQ)</a><a class="headerlink" href="#frequently-asked-questions-faq" title="Permalink to this headline">¶</a></h1>
<ol class="arabic simple">
<li><strong>How can I convert my GTF file into a GFF3 format that MISO accepts?</strong> (<a class="reference external" href="#answer1">answer</a>)</li>
<li><strong>When filtering events from a Bayes factor comparison file, I get the error &#8220;Exception: Malformed comparison line.&#8221;</strong> (<a class="reference external" href="#answer2">answer</a>)</li>
<li><strong>When trying to run MISO on a pickled annotation, I get the error &#8220;ValueError: can&#8217;t handle version -1 of numpy.dtype pickle.&#8221;</strong> (<a class="reference external" href="#answer3">answer</a>)</li>
<li><strong>When I run MISO on my BAM file, it finds that 0 reads align to all events!</strong> (<a class="reference external" href="#answer4">answer</a>)</li>
<li><strong>When I run MISO on my BAM file, it finds far too few reads aligning to the events (but not 0)!</strong> (<a class="reference external" href="#answer5">answer</a>)</li>
<li><strong>I am making my own GFF annotation file of alternative events for use with MISO. How should I format the</strong> <tt class="docutils literal"><span class="pre">ID</span></tt> <strong>value of GFF entries?</strong> (<a class="reference external" href="#answer6">answer</a>)</li>
<li><strong>What are the different MISO versions? Which should I use?</strong> (<a class="reference external" href="#answer7">answer</a>)</li>
<li><strong>Where can I download MISO?</strong> (<a class="reference external" href="#answer8">answer</a>)</li>
<li><strong>How can I map MISO alternative event annotations to the genes they are in?</strong> (<a class="reference external" href="#answer9">answer</a>)</li>
<li><strong>Can I use MISO with my own custom annotations of alternative events? Or my own annotation of genes from their mRNA isoforms that I obtained from Ensembl/UCSC/Refseq?</strong> (<a class="reference external" href="#answer10">answer</a>)</li>
<li><strong>What is the input to the</strong> <tt class="docutils literal"><span class="pre">--compare-samples</span></tt> <strong>option? Do I have to generate a summary file before using it?</strong> (<a class="reference external" href="#answer11">answer</a>)</li>
<li><strong>I ran MISO on genes/events that have 2 isoforms, but the output summary files and comparison files contain just a single number for</strong> Ψ <strong>for each entry. Shouldn&#8217;t there be 2 numbers, one for each isoform?</strong> (<a class="reference external" href="#answer12">answer</a>)</li>
<li><strong>How does MISO deal with replicates?</strong> (<a class="reference external" href="#answer13">answer</a>)</li>
<li><strong>Is there a mapping from events to genes for hg19?</strong>  (<a class="reference external" href="#answer14">answer</a>)</li>
<li><strong>If I am running MISO with a GFF annotation containing multiple isoforms, how do I know which isoform each |Psi| value in the vector corresponds to?</strong> (<a class="reference external" href="#answer15">answer</a>)</li>
<li><strong>When using</strong> <tt class="docutils literal"><span class="pre">pe_utils</span></tt> <strong>to compute the paired-end insert length distribution, I get the error</strong> <tt class="docutils literal"><span class="pre">[main_samview]</span> <span class="pre">truncated</span> <span class="pre">file</span></tt> <strong>and/or a segmentation fault.</strong> (<a class="reference external" href="#answer16">answer</a>)</li>
<li><strong>When I try to index GFF files using</strong> <tt class="docutils literal"><span class="pre">index_gff</span></tt> <strong>I get a segmentation fault.</strong> (<a class="reference external" href="#answer17">answer</a>)</li>
</ol>
<div class="section" id="answers">
<h2><a class="toc-backref" href="#id86">Answers</a><a class="headerlink" href="#answers" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple" id="answer1">
<li><strong>How can I convert my GTF file into a GFF3 format that MISO accepts? Thanks to Ross Lazarus for the answer:</strong> GTF files of gene models can be converted using SequenceOntology&#8217;s <a class="reference external" href="http://www.sequenceontology.org/resources/converter_readme.html">GTF2GFF</a> converter (or using this posted <a class="reference external" href="http://seqanswers.com/forums/showthread.php?t=3201&amp;highlight=GFF3">Perl script</a>) to GFF3 format. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer2" start="2">
<li><strong>When filtering events from a Bayes factor comparison file, I get the error &#8220;Exception: Malformed comparison line.&#8221;</strong> This error arises when using the options to filter on number of inclusion/exclusion reads (or the sum of these) on a comparison file that contains events with more than two isoforms. &#8220;Inclusion&#8221; and &#8220;exclusion&#8221; reads are only defined when the event has two isoforms, where MISO by convention considers the first isoform in the annotation to be the inclusive one, and the second to be the exclusive. Events with three or more isoforms cannot be filtered using these options. In the general case of arbitrarily many isoforms (e.g. some 2 and some more than 2), <tt class="docutils literal"><span class="pre">filter_events.py</span></tt> can be used but then the only valid options are <tt class="docutils literal"><span class="pre">--bayes-factor</span></tt> or <tt class="docutils literal"><span class="pre">--delta-psi</span></tt> to filter on Bayes factor and/or magnitude of change in Ψ, which are defined in all cases. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer3" start="3">
<li><strong>When trying to run MISO on a pickled annotation, I get the error &#8220;ValueError: can&#8217;t handle version -1 of numpy.dtype pickle.&#8221;</strong> This probably means that you are using an older version of numpy. The solution is to use version 1.4 or higher, see Brad Friedman&#8217;s note (<a class="reference external" href="http://blog.friedman-lab.com/2011/01/11/running-miso-on-a-cluster/">http://blog.friedman-lab.com/2011/01/11/running-miso-on-a-cluster/</a>).  Thanks to Brad Friedman and Marvin Jens. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer4" start="4">
<li><strong>When I run MISO on my BAM file, it finds that 0 reads align to all events!</strong> This problem is most commonly caused by either passing MISO a BAM file that is not indexed, or a BAM file that was indexed but where the <tt class="docutils literal"><span class="pre">.bai</span></tt> file (which holds the index) is not present in the same directory as the <tt class="docutils literal"><span class="pre">.bam</span></tt> file passed to MISO. Indexing the file and ensuring that its corresponding index file is in the same directory will fix the problem. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer5" start="5">
<li><strong>When I run MISO on my BAM file, it finds far too few reads aligning to the events (but not 0)!</strong> This is typically caused by passing MISO the wrong <tt class="docutils literal"><span class="pre">--read-len</span></tt> arguments, or by passing MISO files that contain a mix of different length reads. Currently MISO does not support mixed read lengths, though implementation of this feature is in the works. If you pass the argument <tt class="docutils literal"><span class="pre">--read-len</span> <span class="pre">50</span></tt> but the actual reads are of length 52, these reads will not be alignable, and so it is important to check the value of this argument. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer6" start="6">
<li><strong>I am making my own GFF annotation file of alternative events for use with MISO. How should I format the</strong> <tt class="docutils literal"><span class="pre">ID</span></tt> <strong>value of GFF entries?</strong> The GFF <tt class="docutils literal"><span class="pre">ID</span></tt> values do not have to have any particular format and are ignored by MISO. These values are only used to link each entry with its parents/children as usual for the GFF format. As long as the chosen <tt class="docutils literal"><span class="pre">ID</span></tt> values maintain the correct relationship between <tt class="docutils literal"><span class="pre">gene</span></tt>, <tt class="docutils literal"><span class="pre">mRNA</span></tt> and <tt class="docutils literal"><span class="pre">exon</span></tt> entries, they can be of any value. MISO will only use the actual genomic coordinates of each entry when processing the annotation. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer7" start="7">
<li><strong>What are the different MISO versions? Which should I use?</strong> MISO currently has two versions: the original version, written in Python, and a newer version (released October 2011) containing a statistical inference engine written in C &#8211; called <tt class="docutils literal"><span class="pre">fastmiso</span></tt> &#8211; that is accessible through a Python interface. Version <tt class="docutils literal"><span class="pre">fastmiso</span></tt> is roughly 60-100x faster than the Python-only version. The Python interface to <tt class="docutils literal"><span class="pre">fastmiso</span></tt> is identical to the Python interface to the original version of MISO, meaning both take exactly the same inputs and command line arguments, and output files in exactly the same formats. Thus, using the two versions from a user&#8217;s perspective is the same, the only difference is the installation (<tt class="docutils literal"><span class="pre">fastmiso</span></tt> requires compilation using a C compiler like GCC and fewer Python packages to be installaed, while the Python-only version of MISO requires no compilation, but has a larger set of Python package requirements.) We recommend using <tt class="docutils literal"><span class="pre">fastmiso</span></tt>, which now replaces the original Python-only version. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer8" start="8">
<li><strong>Where can I download MISO?</strong> MISO can be downloaded from the <a class="reference internal" href="#releases"><em>Releases</em></a> section (to obtain a stable release), or from our GitHub <a class="reference external" href="https://github.com/yarden/MISO">repository</a> where you can get the latest version. See <a class="reference internal" href="#installation"><em>Installation</em></a> section for more information. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer9" start="9">
<li><strong>How can I map MISO alternative event annotations to the genes they are in?</strong> An easy way to do this is to get the transcription start and end sites for all genes, and overlap these with the MISO alternative event coordinates. We provide our own mapping (based on Ensembl gene models) of alternative events to Ensembl gene IDs for the human and mouse annotations, available here <a class="reference internal" href="#events-to-genes"><em>Mapping of alternative events to genes</em></a>. To create your own mapping of events to genes based on a different annotation, all that is needed is a table containing the transcription start/end sites for all genes. These can be be downloaded from the <a class="reference external" href="http://genome.ucsc.edu/">UCSC Genome Browser</a> (by clicking the &#8220;Tables&#8221; tab), or from Ensembl&#8217;s <a class="reference external" href="http://www.ensembl.org/biomart/martview">BioMart</a>. In UCSC tables, the transcription start/end sites are encoded in the fields <tt class="docutils literal"><span class="pre">txStart</span></tt> or <tt class="docutils literal"><span class="pre">txEnd</span></tt>. You can overlap the coordinates of the events given in our GFF3 fields to these start and end sites in your favorite gene model annotation (e.g. UCSC, Ensembl, or Refseq) and determine what gene they overlap.</li>
</ol>
<ol class="arabic simple" id="answer10" start="10">
<li><strong>Can I use MISO with my own custom annotations of alternative events? Or my own annotation of genes from their mRNA isoforms that I obtained from Ensembl/UCSC/Refseq?</strong> Yes. MISO will take any correctly formatted GFF3 file and quantitate the abundances of the mRNAs listed for each entry in that file. Those entries can represent individual alternative splicing events like our own annotation (yielding an &#8220;exon-centric&#8221; quantitation, as described in the documentation), or they might be the full mRNAs of each gene, as described in a database like Ensembl or UCSC (to perform &#8220;isoform-centric&#8221; quantitation). (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer11" start="11">
<li><strong>What is the input to the</strong> <tt class="docutils literal"><span class="pre">--compare-samples</span></tt> <strong>option? Do I have to generate a summary file before using it?</strong> The input to <tt class="docutils literal"><span class="pre">--compare-samples</span></tt> are two directories (corresponding to the two samples to be compared) that are outputted by a call to <tt class="docutils literal"><span class="pre">--compute-genes-psi</span></tt> (in <tt class="docutils literal"><span class="pre">run_events_analysis.py</span></tt>). There is no need to generate a summary file before calling <tt class="docutils literal"><span class="pre">--compare-samples</span></tt>, and the output of <tt class="docutils literal"><span class="pre">--summarize-samples</span></tt> is not used at all when doing a samples comparison. A summary file is just a single-file summary of the information contained in the directory produced by <tt class="docutils literal"><span class="pre">--compute-genes-psi</span></tt> for a particular sample, made for convenience as a summary of the expression levels of exons/isoforms in that sample. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer12" start="12">
<li><strong>I ran MISO on genes/events that have 2 isoforms, but the output summary files and comparison files contain just a single number for</strong> Ψ <strong>for each entry. Shouldn&#8217;t there be 2 numbers, one for each isoform?</strong> If a gene/event has two isoforms, MISO will quantitate the expression level of the first in the annotation (defined as Ψ), which means that the expression level of the other is simply 1 - Ψ. MISO will always report a comma-separated list of isoforms (listed according to their <tt class="docutils literal"><span class="pre">ID=</span></tt> as given in the input annotation file) in the summary and comparison files for each entry that corresponds to the order of the Ψ values, such that there is no ambiguity about which isoform has the expression level Ψ and which is 1 - Ψ. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer13" start="13">
<li><strong>How does MISO deal with replicates?</strong> There are several ways to deal with alternative splicing expression estimates in replicates, but these are not directly in MISO &#8211; currently, MISO does not handle replicates / groups of samples in any special way.  One approach to replicates is to process the BAM file for each replicate separately and then also run MISO on a pooled version of the replicates (e.g. all the experimental samples and all the controls.) This approach allows you to filter events that change by requiring them to change in the same direction (and additionally with similar orders of magnitude) in all the replicates, or most of the replicates.  The Ψ values for all samples can then be clustered (see Wang et. al. (2008) for example) to determine if the exon/isoform expression level for the replicates in the same condition are more similar to each other than to replicates in different condition. If you find that the biological replicates behave similarly based on a clustering analysis, then it might be possible to use MISO estimates for the pooled samples in downstream analyses, where you&#8217;ll have increased power to detect differences. We find that analysis of the replicates independently first works well, since it can weed out events that show a change in only a minority of the biological replicates if you want to be conservative, but also give you power to be inclusive when needed by analyzing the pooled replicates. It is also possible to frame the replicates problem in a traditional statistical framework: if you have a table of read counts supporting each isoform for an event (this is outputted by MISO &#8211; see <a class="reference internal" href="#usingcounts"><em>Using the read class counts</em></a>) for each replicate in your groups, and then you can see if there are group-wide differences in these counts.  This approach is less direct, but should enable detection of group-wide trends &#8211; in an &#8220;ANOVA-like&#8221; way &#8211; by using established statistical techniques for problems that can be framed as a table of counts across different conditions. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer14" start="14">
<li><strong>Is there a mapping from events to genes for hg19?</strong> Not at the moment. To generate one for your favorite gene models annotation, see <a class="reference external" href="#answer9">Question 9</a> of FAQ. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer15" start="15">
<li><strong>If I am running MISO with a GFF annotation containing multiple isoforms, how do I know which isoform each |Psi| value in the vector corresponds to?</strong> (<a class="reference external" href="#answer14">answer</a>) If given a GFF annotation with multiple (i.e. more than 2) mRNA isoforms, MISO will assign Ψ values to isoforms in the order in which their <tt class="docutils literal"><span class="pre">mRNA</span></tt> entries appear in the GFF file. (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer16" start="16">
<li><strong>When using</strong> <tt class="docutils literal"><span class="pre">pe_utils</span></tt> <strong>to compute the paired-end insert length distribution, I get the error</strong> <tt class="docutils literal"><span class="pre">[main_samview]</span> <span class="pre">truncated</span> <span class="pre">file</span></tt> <strong>and/or a segmentation fault.</strong> This is usually caused because no reads in your BAM file match the GFF file of constitutive exons given as argument to <tt class="docutils literal"><span class="pre">pe_utils</span></tt>, resulting in an empty BAM file that causes downstream steps to fail. The reason is typically a technical mismatch between the BAM and GFF, like different chromosome naming conventions (e.g. <tt class="docutils literal"><span class="pre">chr</span></tt> prefix chromosome names in the BAM but not in the constitutive exons GFF). See <a class="reference internal" href="#iso-centric"><em>Human/mouse gene models for isoform-centric analyses</em></a> for more information.  (<a class="reference external" href="#faq">back</a>)</li>
</ol>
<ol class="arabic simple" id="answer17" start="17">
<li><strong>When I try to index GFF files using</strong> <tt class="docutils literal"><span class="pre">index_gff</span></tt> <strong>I get a segmentation fault.</strong> The indexing step requires the <tt class="docutils literal"><span class="pre">shelve</span></tt> module from Python. On some systems and versions of Python, the <tt class="docutils literal"><span class="pre">shelve</span></tt> module can cause a segmentation fault. This is likely a Python installation problem that should be resolved locally. Any Python script that makes use of the <tt class="docutils literal"><span class="pre">shelve.open</span></tt> function will be affected.</li>
</ol>
</div>
</div>
<div class="section" id="single-end-event-counts-sec-format">
<span id="sec"></span><h1><a class="toc-backref" href="#id87">Single-end Event Counts (SEC) format</a><a class="headerlink" href="#single-end-event-counts-sec-format" title="Permalink to this headline">¶</a></h1>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The SEC format is deprecated - we recommend using BAM files along with GFF annotation files instead, as that format is more flexible and standard, and allows representation of both individual alternative splicing events (for exon-centric) and whole mRNA isoforms (for isoform-centric analyses).</p>
</div>
<p>The Single-end Event Counts (SEC) format can be used to quantitate individual alternative splicing events
(exon-centric analyses) using single-end RNA-Seq data sets. This format does not support paired-end
information and is best suited for cases where individual events are quantitated. The format currently
supports skipped exons (SE), alternative 3&#8217; UTRs (TandemUTR), and retained introns (RI). The format is
described below.</p>
<div class="section" id="sec-format-specification">
<h2><a class="toc-backref" href="#id88">SEC format specification</a><a class="headerlink" href="#sec-format-specification" title="Permalink to this headline">¶</a></h2>
<p>The SEC format is a two-column <em>tab-separated</em> format, where the first column identifies the coordinates of the
alternative events, and the second gives a set of counts. The elements in each field as always
separated by semicolons (<tt class="docutils literal"><span class="pre">;</span></tt>).</p>
<div class="section" id="skipped-exons-se-format">
<h3><a class="toc-backref" href="#id89">Skipped exons (SE) format</a><a class="headerlink" href="#skipped-exons-se-format" title="Permalink to this headline">¶</a></h3>
<p>For skipped exons, the SEC format is:</p>
<div class="highlight-python"><pre>up_exon;skipped_exon;down_exon        up;se;dn;ijxn1;ijxn2;ejxn</pre>
</div>
<p>where the first column specifies the coordinates as follows:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">up_exon</span></tt>: coordinates of the exon upstream (5&#8217;) of the skipped exon,</li>
<li><tt class="docutils literal"><span class="pre">skipped_exon</span></tt>: coordinates of the skipped exon,</li>
<li><tt class="docutils literal"><span class="pre">down_exon</span></tt>: coordinates of the exon downstream (3&#8217;) of the skipped exon</li>
</ul>
<p>and the second column specifies the counts for the skipped exon event:</p>
<blockquote>
<div><a class="reference internal image-reference" href="_images/MISO-SEC-SE.png"><img alt="SEC format for SE" class="align-right" src="_images/MISO-SEC-SE.png" style="width: 265.6px; height: 259.2px;" /></a>
</div></blockquote>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">up</span></tt>: number of reads aligning to the upstream exon body,</li>
<li><tt class="docutils literal"><span class="pre">se</span></tt>: number of reads aligning to the skipped exon body,</li>
<li><tt class="docutils literal"><span class="pre">dn</span></tt>: number of reads aligning to the downstream exon body,</li>
<li><tt class="docutils literal"><span class="pre">ijxn1</span></tt>: number of reads aligning to the &#8220;upstream inclusion junction&#8221; (connecting the upstream and skipped exons)</li>
<li><tt class="docutils literal"><span class="pre">ijxn2</span></tt>: number of reads aligning to the &#8220;downstream inclusion junction&#8221; (connecting the skipped and downstream exons)</li>
<li><tt class="docutils literal"><span class="pre">ejxn</span></tt>: number of reads aligning to the &#8220;exclusion junction&#8221; (connecting the upstream and downstream exons)</li>
</ul>
<p>The reads corresponding to each set of counts are shown in the figure above.</p>
<p>An example SE event is:</p>
<div class="highlight-python"><pre>chr1:88443357:88443232:-;chr1:88442382:88442277:-;chr1:88440158:88439569:-    22;39;91;2;2;1</pre>
</div>
</div>
<div class="section" id="alternative-tandem-3-utrs-tandemutr-format">
<h3><a class="toc-backref" href="#id90">Alternative tandem 3&#8217; UTRs (TandemUTR) format</a><a class="headerlink" href="#alternative-tandem-3-utrs-tandemutr-format" title="Permalink to this headline">¶</a></h3>
<p>For alternative tandem 3&#8217; UTRs, the SEC format is:</p>
<div class="highlight-python"><pre>core_coords;ext_coords        ext_count;core_count</pre>
</div>
<p>where the first column specifies the coordinates of the core and extension regions of the TandemUTR event as follows:</p>
<blockquote>
<div><a class="reference internal image-reference" href="_images/MISO-SEC-TandemUTR.png"><img alt="SEC format for TandemUTR" class="align-right" src="_images/MISO-SEC-TandemUTR.png" style="width: 366.0px; height: 156.4px;" /></a>
</div></blockquote>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">core_coords</span></tt>: coordinates of the core region,</li>
<li><tt class="docutils literal"><span class="pre">ext_coords</span></tt>: coordinates of the extension region</li>
</ul>
<p>and the second column specifies the counts in the regions:</p>
<ul>
<li><p class="first"><tt class="docutils literal"><span class="pre">ext_count</span></tt>: number of reads in the extension region,</p>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">core_count</span></tt>: number of reads in the core region</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The coordinates of the event are given in the order core, extension but the counts are given in the <em>opposite order</em> of extension count followed by core count.</p>
</div>
</li>
</ul>
<p>The reads corresponding to each count are shown in the figure above.</p>
<p>An example TandemUTR event is:</p>
<div class="highlight-python"><pre>chr1:136527951:136528770:+;chr1:136528771:136529456:+  41;147</pre>
</div>
</div>
<div class="section" id="retained-intron-ri-format">
<h3><a class="toc-backref" href="#id91">Retained intron (RI) format</a><a class="headerlink" href="#retained-intron-ri-format" title="Permalink to this headline">¶</a></h3>
<p>For retained introns, the SEC format is:</p>
<div class="highlight-python"><pre>up_exon;retained_intron;dn_exon       up;ri;dn;ejxn</pre>
</div>
<p>where the first column specifies the coordinates of the retained intron and its flanking exons as follows:</p>
<blockquote>
<div><a class="reference internal image-reference" href="_images/MISO-SEC-RI.png"><img alt="SEC format for RI" class="align-right" src="_images/MISO-SEC-RI.png" style="width: 275.6px; height: 247.6px;" /></a>
</div></blockquote>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">up_exon</span></tt>: coordinates of the exon upstream (5&#8217;) of the retained intron</li>
<li><tt class="docutils literal"><span class="pre">retained_intron</span></tt>: coordinates of the retained intron</li>
<li><tt class="docutils literal"><span class="pre">dn_exon</span></tt>: coordinates of the exon downstream (3&#8217;) of the retained intron</li>
</ul>
<p>and the second column specifies the counts in the regions:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">up</span></tt>: number of reads in the upstream exon,</li>
<li><tt class="docutils literal"><span class="pre">ri</span></tt>: number of reads in the retained intron. This includes: (1) reads in the junction between the upstream exon and retained intron, (2) reads in the body of the retained intron, and (3) reads in the junction between the retained intron and downstream exon,</li>
<li><tt class="docutils literal"><span class="pre">dn</span></tt>: number of reads in the downstream exon,</li>
<li><tt class="docutils literal"><span class="pre">ejxn</span></tt>: number of reads in the &#8220;exclusion junction&#8221; connecting the upstream and downstream exon (these reads support splicing out of the intron.)</li>
</ul>
<p>An example RI event is:</p>
<div class="highlight-python"><pre>chr3:141907860:141907975:-;chr3:141907976:141910258:-;chr3:141910259:141910379:-      171;8;281;103</pre>
</div>
<p>The reads corresponding to each count are shown in the figure above.</p>
</div>
<div class="section" id="running-miso-on-sec-files">
<h3><a class="toc-backref" href="#id92">Running MISO on SEC files</a><a class="headerlink" href="#running-miso-on-sec-files" title="Permalink to this headline">¶</a></h3>
<p>To run MISO on a SEC format, the <tt class="docutils literal"><span class="pre">run_events_analysis.py</span> <span class="pre">--compute-events-psi</span></tt> script is used:</p>
<div class="highlight-python"><pre>--compute-events-psi
                      Compute Psi values for all events. Expects two
                      arguments: a set of labels and a set of filenames with
                      associated read counts.</pre>
</div>
<p>The first argument to <tt class="docutils literal"><span class="pre">--compute-events-psi</span></tt> is a comma-separated set of sample labels and the second is a comma-separated set of counts filenames. For instance, <tt class="docutils literal"><span class="pre">--compute-events-psi</span> <span class="pre">control,knockdown</span> <span class="pre">control.counts,knockdown.counts</span></tt> would run MISO on the SEC files <tt class="docutils literal"><span class="pre">control.counts</span></tt> and <tt class="docutils literal"><span class="pre">knockdown.counts</span></tt>, labeling the output directories <tt class="docutils literal"><span class="pre">control</span></tt> and <tt class="docutils literal"><span class="pre">knockdown</span></tt>, respectively. When running on a single sample, the comma notation can be dropped (e.g. <tt class="docutils literal"><span class="pre">--compute-events-psi</span> <span class="pre">control</span> <span class="pre">control.counts</span></tt>).</p>
<p>To use <tt class="docutils literal"><span class="pre">--compute-events-psi</span></tt>, the type of event must be specified through the <tt class="docutils literal"><span class="pre">--event-type</span></tt> option, which
can be either <tt class="docutils literal"><span class="pre">SE</span></tt> (skipped exons), <tt class="docutils literal"><span class="pre">TandemUTR</span></tt> (alt. tandem UTRs) or <tt class="docutils literal"><span class="pre">RI</span></tt> (retained introns.) The sequenced
read length must be specified with <tt class="docutils literal"><span class="pre">--read-len</span></tt>. Optionally, the overhang length imposed on junction reads during alignment can be specified with <tt class="docutils literal"><span class="pre">--overhang-len</span></tt>.</p>
<p>For example, the following command would run MISO on a set of skipped exons in the SEC file <tt class="docutils literal"><span class="pre">control.counts</span></tt> in an experiment where the reads are 35 nt long and an overhang length of 4 was placed on junction reads during alignment:</p>
<div class="highlight-python"><pre>python run_events_analysis.py --compute-events-psi control control.counts --event-type SE --read-len 35 --overhang-len 4 --output-dir control</pre>
</div>
<p>This will create a directory called <tt class="docutils literal"><span class="pre">SE/control/</span></tt> with MISO output for each event in <tt class="docutils literal"><span class="pre">control.counts</span></tt> that meets the default read counts filter.</p>
</div>
</div>
</div>
<div class="section" id="advanced-uses-of-miso">
<h1><a class="toc-backref" href="#id93">Advanced uses of MISO</a><a class="headerlink" href="#advanced-uses-of-miso" title="Permalink to this headline">¶</a></h1>
<p>Rather than estimating the expression level of a single alternatively spliced exon (&#8220;exon-centric&#8221;),
or of each transcript belonging to a gene (&#8220;isoform-centric&#8221;), one can quantify the levels of multiple
isoforms produced by several nearby alternative splicing events. This is called
<em>local multi-isoform quantitation</em>. These isoforms can be encoded in the GFF format as for the
case of alternative splicing events that produce two isoforms, described in <a class="reference internal" href="#gff-event-annotation"><em>GFF-based alternative events format</em></a>.</p>
<blockquote>
<div><div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Section under construction</p>
</div>
</div></blockquote>
</div>
<div class="section" id="acknowledgements">
<span id="katz"></span><span id="refs"></span><h1><a class="toc-backref" href="#id94">Acknowledgements</a><a class="headerlink" href="#acknowledgements" title="Permalink to this headline">¶</a></h1>
<p>Thanks to the following users for their feedback/comments on MISO and
for suggesting many improvements, or for creating useful software used by MISO:</p>
<ul class="simple">
<li>Sol Katzman (UCSC)</li>
<li>Brenton Graveley (UConn)</li>
<li>Michael Lovci (UCSD)</li>
<li>Marvin Jens (Max-Delbrueck-Center for Molecular Medicine)</li>
<li>Schragi Schwartz (Broad Institute)</li>
<li>Rahul Satija (Broad Institute)</li>
<li>Henry Bigelow (Broad Institute)</li>
<li>Ersen Kavak (Karolinska Institute)</li>
<li>Essi Laajala (University of Turku)</li>
<li>Kuan-Ting Lin (CSHL)</li>
<li>Aaron Quinlan (University of Virginia)</li>
<li>Ross Lazarus (Harvard Medical School)</li>
<li>Isaac Berkowitz (Brown University)</li>
<li>Chia-Ho Lin (UCLA)</li>
<li>Schragi Schwartz (MIT)</li>
<li>Shaun Mahony (MIT)</li>
<li>Albert W. Cheng (MIT)</li>
<li>Rory Kirchner (MIT)</li>
<li>Robert K. Bradley (MIT)</li>
<li>Brad Friedman (MIT)</li>
<li>Vincent Butty (MIT)</li>
<li>Jessica Hurt (MIT)</li>
<li>Reut Shalgi (MIT)</li>
<li>Colin Dewey (University of Wisconsin-Madison)</li>
<li>Eric Suh (Princeton)</li>
</ul>
</div>
<div class="section" id="authors">
<h1><a class="toc-backref" href="#id95">Authors</a><a class="headerlink" href="#authors" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li>The original Python version of MISO was written by Yarden Katz. The C version of MISO was written by Gábor Csárdi, with Python interfaces written by Yarden Katz and an R interface written by Gábor Csárdi. Parts of the code for visualizing MISO estimates was written by Eric T. Wang. The MISO manual was written by Yarden Katz.</li>
</ul>
</div>
<div class="section" id="installing-python-only-miso-version">
<span id="installing-pymiso"></span><h1><a class="toc-backref" href="#id96">Installing Python-only MISO version</a><a class="headerlink" href="#installing-python-only-miso-version" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The Python-only version of MISO is now deprecated &#8211; we recommend using <tt class="docutils literal"><span class="pre">fastmiso</span></tt>.</p>
</div>
</div></blockquote>
<p>This version requires:</p>
<ol class="arabic simple">
<li><a class="reference external" href="http://www.python.org">Python 2.6</a> or higher</li>
<li><a class="reference external" href="http://www.numpy.org">numpy</a> / <a class="reference external" href="http://www.scipy.org">scipy</a>. <strong>NOTE</strong>: MISO requires numpy version &gt; 1.4!</li>
<li><a class="reference external" href="http://matplotlib.sourceforge.net/">matplotlib</a></li>
<li><a class="reference external" href="http://samtools.sourceforge.net/">samtools</a> for accessing SAM/BAM files</li>
<li><a class="reference external" href="http://code.google.com/p/pysam/">pysam</a>, a Python library for working with SAM/BAM files through <tt class="docutils literal"><span class="pre">samtools</span></tt> (<strong>NOTE</strong>: MISO requires pysam version 0.6 or higher)</li>
</ol>
<p>MISO has not been tested with Python 3.x. Once these libraries are available, the Python-only version of MISO can be used.</p>
<p>To get the most up to date version of the Python-only MISO, we recommend checking out the latest version of the <tt class="docutils literal"><span class="pre">dev</span></tt> branch from our GitHub <a class="reference external" href="https://github.com/yarden/MISO">repository</a>.</p>
</div>
<div class="section" id="references">
<span id="myrefs"></span><h1><a class="toc-backref" href="#id97">References</a><a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h1>
<ol class="arabic simple">
<li>Katz Y, Wang ET, Airoldi EM, Burge CB. (2010). <a class="reference external" href="http://www.nature.com/nmeth/journal/v7/n12/full/nmeth.1528.html">Analysis and design of RNA sequencing experiments for identifying isoform regulation</a>. Nature Methods 7, 1009-1015.</li>
<li>Wang ET, Sandberg R, Luo S, Khrebtukova I, Zhang L, Mayr C, Kingsmore SF, Schroth GP, Burge CB. (2008). Alternative Isoform Regulation in Human Tissue Transcriptomes. Nature 456, 470-476</li>
</ol>
<div class="section" id="related-software">
<h2><a class="toc-backref" href="#id98">Related software</a><a class="headerlink" href="#related-software" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference external" href="sashimi.html">sashimi_plot</a>: tool for visualizing RNA-Seq data and plotting MISO output</li>
<li><a class="reference external" href="http://bowtie-bio.sourceforge.net/">Bowtie</a>: for alignment of RNA-Seq reads to genome (optionally with precomputed set of splice junctions)</li>
<li><a class="reference external" href="http://tophat.cbcb.umd.edu/">Tophat</a>: for spliced alignment of RNA-Seq reads to genome with <em>de
novo</em> splice junction discovery</li>
<li><a class="reference external" href="http://www.broadinstitute.org/igv/">IGV</a>: Visualizer of mapped reads (e.g. BAM files). Displays junction reads.</li>
<li><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a>: another isoform expression/differential regulation software with transcript assembly</li>
</ul>
</div>
<div class="section" id="related-reading">
<h2><a class="toc-backref" href="#id99">Related reading</a><a class="headerlink" href="#related-reading" title="Permalink to this headline">¶</a></h2>
<div class="section" id="probabilistic-models-in-computational-biology">
<h3><a class="toc-backref" href="#id100">Probabilistic models in computational biology</a><a class="headerlink" href="#probabilistic-models-in-computational-biology" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Airoldi, E.M. (2007). <a class="reference external" href="http://dx.doi.org/10.1371/journal.pcbi.0030252">Getting started in probabilistic graphical models</a>. PLoS Computational Biology, 3, e252.</li>
</ul>
</div>
<div class="section" id="bayes-factors-and-bayesian-hypothesis-testing">
<h3><a class="toc-backref" href="#id101">Bayes factors and Bayesian hypothesis testing</a><a class="headerlink" href="#bayes-factors-and-bayesian-hypothesis-testing" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><a class="reference external" href="http://www.socsci.uci.edu/~mdlee/bgm.html">A Practical Course in Bayesian Graphical Modeling</a> by Michael Lee and Eric-Jan Wagenmakers: Free online book that contains (among many other things) excellent chapters on Bayesian hypothesis testing with practical examples, including discussion of Bayes factors and the Savage-Dickey ratio test.</li>
</ul>
</div>
<div class="section" id="general-reading-on-probabilistic-modeling-and-inference">
<h3><a class="toc-backref" href="#id102">General reading on probabilistic modeling and inference</a><a class="headerlink" href="#general-reading-on-probabilistic-modeling-and-inference" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Griffiths, T. L., &amp; Yuille, A. (2006). <a class="reference external" href="http://cocosci.berkeley.edu/tom/papers/tutorial.pdf">A primer on probabilistic inference</a>. Trends in Cognitive Sciences, 10.</li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">What is MISO?</a></li>
<li><a class="reference internal" href="#how-miso-works">How MISO works</a></li>
<li><a class="reference internal" href="#features">Features</a></li>
<li><a class="reference internal" href="#mailing-list">Mailing list</a></li>
<li><a class="reference internal" href="#updates">Updates</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#releases">Releases</a></li>
<li><a class="reference internal" href="#latest-version-from-github">Latest version from GitHub</a></li>
<li><a class="reference internal" href="#installation-requirements">Installation requirements</a></li>
<li><a class="reference internal" href="#quickstart">Quickstart</a></li>
<li><a class="reference internal" href="#installing-the-fastmiso-version-of-miso">Installing the <tt class="docutils literal"><span class="pre">fastmiso</span></tt> version of MISO</a><ul>
<li><a class="reference internal" href="#installing-using-a-stable-release">Installing using a stable release</a></li>
<li><a class="reference internal" href="#installing-the-latest-version-from-github-repository">Installing the latest version from GitHub repository</a></li>
</ul>
</li>
<li><a class="reference internal" href="#testing-the-installation">Testing the installation</a></li>
<li><a class="reference internal" href="#testing-miso">Testing MISO</a></li>
</ul>
</li>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#ways-of-running-miso-and-associated-file-formats">Ways of running MISO and associated file formats</a></li>
<li><a class="reference internal" href="#using-miso-on-a-cluster">Using MISO on a cluster</a></li>
<li><a class="reference internal" href="#alternative-event-annotations">Alternative event annotations</a><ul>
<li><a class="reference internal" href="#human-mouse-fly-alternative-event-annotations-files">Human/mouse/fly alternative event annotations files</a></li>
<li><a class="reference internal" href="#gff-based-alternative-events-format">GFF-based alternative events format</a></li>
<li><a class="reference internal" href="#mapping-of-alternative-events-to-genes">Mapping of alternative events to genes</a></li>
<li><a class="reference internal" href="#using-gff-event-annotations-with-paired-end-reads">Using GFF event annotations with paired-end reads</a></li>
<li><a class="reference internal" href="#human-mouse-gene-models-for-isoform-centric-analyses">Human/mouse gene models for isoform-centric analyses</a></li>
</ul>
</li>
<li><a class="reference internal" href="#running-miso">Running MISO</a><ul>
<li><a class="reference internal" href="#invoking-miso-from-the-command-line">Invoking MISO from the command line</a></li>
<li><a class="reference internal" href="#preparing-the-rna-seq-reads">Preparing the RNA-Seq reads</a></li>
<li><a class="reference internal" href="#preparing-the-alternative-isoforms-annotation">Preparing the alternative isoforms annotation</a></li>
<li><a class="reference internal" href="#configuring-miso">Configuring MISO</a></li>
<li><a class="reference internal" href="#computing-expression-estimates-psi-psi-values">Computing expression estimates (&#8220;Psi&#8221; / Ψ values)</a><ul>
<li><a class="reference internal" href="#using-single-end-reads">Using single-end reads</a></li>
<li><a class="reference internal" href="#using-paired-end-reads">Using paired-end reads</a><ul>
<li><a class="reference internal" href="#computing-the-insert-length-distribution-and-its-statistics">Computing the insert length distribution and its statistics</a><ul>
<li><a class="reference internal" href="#getting-a-set-of-long-constitutive-exons">Getting a set of long constitutive exons</a></li>
<li><a class="reference internal" href="#computing-the-insert-length-distribution-given-a-constitutive-exons-file">Computing the insert length distribution given a constitutive exons file</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#distributing-miso-jobs-to-a-cluster">Distributing MISO jobs to a cluster</a></li>
<li><a class="reference internal" href="#raw-miso-output-format">Raw MISO output format</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#summarizing-miso-output">Summarizing MISO output</a><ul>
<li><a class="reference internal" href="#summary-file-output-format">Summary file output format</a></li>
</ul>
</li>
<li><a class="reference internal" href="#detecting-differentially-expressed-isoforms">Detecting differentially expressed isoforms</a><ul>
<li><a class="reference internal" href="#output-of-samples-comparison">Output of samples comparison</a><ul>
<li><a class="reference internal" href="#bayes-factors-file-output-format">Bayes factors file output format</a></li>
<li><a class="reference internal" href="#raw-delta-psi-distribution-output-format">Raw Δ Ψ distribution output format</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#example-miso-pipeline">Example MISO pipeline</a></li>
<li><a class="reference internal" href="#interpreting-and-filtering-miso-output">Interpreting and filtering MISO output</a><ul>
<li><a class="reference internal" href="#using-the-read-class-counts">Using the read class counts</a></li>
<li><a class="reference internal" href="#using-the-confidence-intervals">Using the confidence intervals</a></li>
<li><a class="reference internal" href="#filtering-differentially-expressed-events">Filtering differentially expressed events</a></li>
</ul>
</li>
<li><a class="reference internal" href="#visualizing-and-plotting-miso-output">Visualizing and plotting MISO output</a></li>
<li><a class="reference internal" href="#frequently-asked-questions-faq">Frequently Asked Questions (FAQ)</a><ul>
<li><a class="reference internal" href="#answers">Answers</a></li>
</ul>
</li>
<li><a class="reference internal" href="#single-end-event-counts-sec-format">Single-end Event Counts (SEC) format</a><ul>
<li><a class="reference internal" href="#sec-format-specification">SEC format specification</a><ul>
<li><a class="reference internal" href="#skipped-exons-se-format">Skipped exons (SE) format</a></li>
<li><a class="reference internal" href="#alternative-tandem-3-utrs-tandemutr-format">Alternative tandem 3&#8217; UTRs (TandemUTR) format</a></li>
<li><a class="reference internal" href="#retained-intron-ri-format">Retained intron (RI) format</a></li>
<li><a class="reference internal" href="#running-miso-on-sec-files">Running MISO on SEC files</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#advanced-uses-of-miso">Advanced uses of MISO</a></li>
<li><a class="reference internal" href="#acknowledgements">Acknowledgements</a></li>
<li><a class="reference internal" href="#authors">Authors</a></li>
<li><a class="reference internal" href="#installing-python-only-miso-version">Installing Python-only MISO version</a></li>
<li><a class="reference internal" href="#references">References</a><ul>
<li><a class="reference internal" href="#related-software">Related software</a></li>
<li><a class="reference internal" href="#related-reading">Related reading</a><ul>
<li><a class="reference internal" href="#probabilistic-models-in-computational-biology">Probabilistic models in computational biology</a></li>
<li><a class="reference internal" href="#bayes-factors-and-bayesian-hypothesis-testing">Bayes factors and Bayesian hypothesis testing</a></li>
<li><a class="reference internal" href="#general-reading-on-probabilistic-modeling-and-inference">General reading on probabilistic modeling and inference</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/index.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li><a href="#">MISO v1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2010, Yarden Katz, Gábor Csárdi, Eric T. Wang, Edoardo M. Airoldi, Christopher B. Burge.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.7.
    </div>
  </body>
</html>