

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>What is sashimi_plot? &mdash; MISO v1.0 documentation</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="MISO v1.0 documentation" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="index.html">MISO v1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <a class="reference internal image-reference" href="_images/sashimi-plot.png"><img alt="sashimi_plot: a tool for visualizing raw RNA-Seq data and MISO output" class="align-right" src="_images/sashimi-plot.png" style="width: 494.55px; height: 287.55px;" /></a>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#what-is-sashimi-plot" id="id2">What is sashimi_plot?</a></li>
<li><a class="reference internal" href="#features" id="id3">Features</a></li>
<li><a class="reference internal" href="#why-is-it-called-sashimi-plot" id="id4">Why is it called sashimi_plot?</a></li>
<li><a class="reference internal" href="#updates" id="id5">Updates</a></li>
<li><a class="reference internal" href="#installation" id="id6">Installation</a><ul>
<li><a class="reference internal" href="#testing-sashimi-plot" id="id7">Testing sashimi_plot</a></li>
</ul>
</li>
<li><a class="reference internal" href="#overview" id="id8">Overview</a><ul>
<li><a class="reference internal" href="#visualizing-rna-seq-reads-along-isoforms-with-miso-estimates" id="id9">Visualizing RNA-Seq reads along isoforms with MISO estimates</a></li>
<li><a class="reference internal" href="#command-line-options" id="id10">Command-line options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#visualizing-and-plotting-miso-output" id="id11">Visualizing and plotting MISO output</a><ul>
<li><a class="reference internal" href="#customizing-the-appearance-of-miso-estimates-alongside-raw-rna-seq-data" id="id12">Customizing the appearance of MISO estimates alongside raw RNA-Seq data</a></li>
<li><a class="reference internal" href="#plotting-the-distribution-of-events-with-bayes-factors" id="id13">Plotting the distribution of events with Bayes factors</a></li>
<li><a class="reference internal" href="#plotting-miso-estimates-individually" id="id14">Plotting MISO estimates individually</a></li>
<li><a class="reference internal" href="#plotting-insert-length-distributions-and-summary-statistics" id="id15">Plotting insert length distributions and summary statistics</a></li>
</ul>
</li>
<li><a class="reference internal" href="#frequently-asked-questions-faq" id="id16">Frequently Asked Questions (FAQ)</a></li>
<li><a class="reference internal" href="#authors" id="id17">Authors</a></li>
<li><a class="reference internal" href="#acknowledgements" id="id18">Acknowledgements</a></li>
<li><a class="reference internal" href="#references" id="id19">References</a><ul>
<li><a class="reference internal" href="#related-software" id="id20">Related software</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="what-is-sashimi-plot">
<h1><a class="toc-backref" href="#id2">What is sashimi_plot?</a><a class="headerlink" href="#what-is-sashimi-plot" title="Permalink to this headline">¶</a></h1>
<p><tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> is a utility for automatically producing publication-quality plots for RNA-Seq analyses of isoform expression. It is part of the <a class="reference external" href="http://genes.mit.edu/burgelab/miso/">MISO</a> framework. In particular, <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> can: (1) plot raw RNA-Seq densities along exons and junctions for multiple samples, while simultaneously visualizing the gene model/isoforms to which reads map, and (2) plot MISO output alongside the raw data or separately.</p>
<p>The MISO framework is described in Katz et. al., <a class="reference external" href="http://www.nature.com/nmeth/journal/v7/n12/full/nmeth.1528.html">Analysis and design of RNA sequencing experiments for identifying isoform regulation</a>. <em>Nature Methods</em> (2010).</p>
<blockquote>
<div><div class="align-center figure align-center">
<a class="reference internal image-reference" href="_images/sashimi-plot-example.png"><img alt="An example plot made by sashimi_plot" src="_images/sashimi-plot-example.png" style="width: 809.6px; height: 546.4px;" /></a>
<p class="caption"><em>An example figure produced by</em> <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt></p>
</div>
</div></blockquote>
</div>
<div class="section" id="features">
<h1><a class="toc-backref" href="#id3">Features</a><a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li>Plots RNA-Seq read densities along exons and junctions, as well as
visualizes the structure of the gene&#8217;s isoforms</li>
<li>Plots of MISO estimates (including full distribution and/or
confidence intervals) for the events in question, showing the
estimates for multiple samples in parallel</li>
<li>Plots insert length distributions for paired-end RNA-Seq samples</li>
<li>Allows visualization of multiple samples on the same figure</li>
<li>Generates publication-quality figures in a variety of flexible
formats (including PDF and PNG)</li>
</ul>
</div>
<div class="section" id="why-is-it-called-sashimi-plot">
<h1><a class="toc-backref" href="#id4">Why is it called sashimi_plot?</a><a class="headerlink" href="#why-is-it-called-sashimi-plot" title="Permalink to this headline">¶</a></h1>
<p>We chose &#8220;Sashimi&#8221; because our tool plots the <em>raw</em> RNA-Seq data in addition to inferences made about the RNA-Seq reads (hat tip to Vincent Butty.) Also, the variations and various &#8220;bumps&#8221; in read densities that one commonly observes also look a bit like rolls of Sashimi. Besides, we thought Sashimi would go well with <a class="reference external" href="http://genes.mit.edu/burgelab/miso/">MISO</a>.</p>
</div>
<div class="section" id="updates">
<h1><a class="toc-backref" href="#id5">Updates</a><a class="headerlink" href="#updates" title="Permalink to this headline">¶</a></h1>
<p><strong>2012</strong></p>
<ul class="simple">
<li><strong>Wed, Feb 1</strong>: New features:</li>
</ul>
<blockquote>
<div><ul class="simple">
<li>Dynamic but consistent y-axis scaling: if <tt class="docutils literal"><span class="pre">ymax</span></tt> is omitted, the same yscale will be chosen for all samples, whose maximum value is determined by the maximum y-value across all samples being plotted.</li>
<li>x-axis for <tt class="docutils literal"><span class="pre">bar_posterior</span></tt> feature is now thinner and has smaller ticks, which results in much better visualization (especially if you&#8217;re plotting many samples.)</li>
</ul>
</div></blockquote>
<ul class="simple">
<li><strong>Sun, Jan 8</strong>: Several changes:</li>
</ul>
<blockquote>
<div><ul class="simple">
<li>Bugfix for scenario that caused plots to fail with <tt class="docutils literal"><span class="pre">KeyError</span></tt> on some events</li>
<li>Now skipping reads with insertions or deletions in their CIGAR strings</li>
<li>Figure height/width now correctly read from settings file</li>
<li>Optional <tt class="docutils literal"><span class="pre">sample_labels</span></tt> argument for labeling each sample&#8217;s track in <tt class="docutils literal"><span class="pre">--plot-event</span></tt></li>
</ul>
</div></blockquote>
<p>Thanks to Sol Katzman, Michael Lovci, Sean O&#8217;Keeffe and Vincent Butty for their contributions and suggestions.</p>
<ul class="simple">
<li><strong>Tue, Jan 3</strong>: Added feature for plotting the distribution Bayes factors (<tt class="docutils literal"><span class="pre">--plot-bf-dist</span></tt>). Note that this feature is only available in <tt class="docutils literal"><span class="pre">misopy-0.2</span></tt> and higher.</li>
</ul>
<p><strong>2011</strong></p>
<ul class="simple">
<li><strong>Mon, Dec 26</strong>: We&#8217;ve reorganized the codebase to be a proper Python module that is part of MISO. This forced us to change the name from <tt class="docutils literal"><span class="pre">sashimi-plot</span></tt> to <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> (underscore now instead of dash), so please change your code accordingly. <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> can now be imported as a module (using <tt class="docutils literal"><span class="pre">misopy.sashimi_plot</span></tt>) and no longer relies on MISO being explicitly in your path. We also fixed some issues with CIGAR string parsing and fixed other minor bugs.</li>
<li><strong>Monday, December 19</strong>: Added support for all event types from MISO annotation. Previous versions could not handle ALE/AFE events properly.</li>
<li><strong>Sunday, December 18</strong>: <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> now recursively searches the subdirectories of paths given in the <tt class="docutils literal"><span class="pre">miso_files</span></tt> section of the configuration file to find the <tt class="docutils literal"><span class="pre">.miso</span></tt> file associated with an event. This should allow more flexibility in organization of MISO output directories that the plotting program recognizes.</li>
<li><strong>Saturday, December 3 (12/2/11)</strong>: <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> is released!</li>
</ul>
</div>
<div class="section" id="installation">
<h1><a class="toc-backref" href="#id6">Installation</a><a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p><tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> comes packaged with <a class="reference external" href="http://genes.mit.edu/burgelab/miso/">MISO</a>. If you have MISO installed, you already have <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt>.
<tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> itself only requires the Python package <a class="reference external" href="http://matplotlib.sourceforge.net/">matplotlib</a> (version 1.1.0 or higher) as well as <a class="reference external" href="http://samtools.sourceforge.net/">samtools</a>.</p>
<div class="section" id="testing-sashimi-plot">
<h2><a class="toc-backref" href="#id7">Testing sashimi_plot</a><a class="headerlink" href="#testing-sashimi-plot" title="Permalink to this headline">¶</a></h2>
<p id="id1">To test that <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> is working, we first need to get a GFF annotation of the alternative events to be visualized. An example GFF annotation file of events is provided in <tt class="docutils literal"><span class="pre">sashimi_plot/test-data/events.gff</span></tt>. This GFF file has to be indexed in order to be used with MISO, with the <tt class="docutils literal"><span class="pre">index_gff.py</span></tt> script:</p>
<div class="highlight-python"><pre>cd sashimi_plot/
python ../index_gff.py --index test-data/events.gff test-data/event-data/</pre>
</div>
<p>We can now plot this event by running the following from within the <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> directory in MISO:</p>
<div class="highlight-python"><pre>python plot.py --plot-event "chr17:45816186:45816265:-@chr17:45815912:45815950:-@chr17:45814875:45814965:-" test-data/event-data/ settings/sashimi_plot_settings.txt --output-dir test-plot</pre>
</div>
<p>If successful, you should get a plot in the directory <tt class="docutils literal"><span class="pre">test-plot/</span></tt> called <tt class="docutils literal"><span class="pre">chr17:45816186:45816265:-&#64;chr17:45815912:45815950:-&#64;chr17:45814875:45814965:-.pdf</span></tt>. An annotated graphical explanation of the main features of the output is shown below.</p>
<blockquote>
<div><div class="align-center figure align-center">
<a class="reference internal image-reference" href="_images/sashimi-plot-example-annotated.png"><img alt="sashimi_plot example with annotations" src="_images/sashimi-plot-example-annotated.png" style="width: 992.95px; height: 485.45px;" /></a>
<p class="caption"><tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> <em>example figure with annotations</em></p>
</div>
</div></blockquote>
<p>Key items to notice:</p>
<ul class="simple">
<li>The RNA-Seq read densities along exons are shown as histograms, color-coded by the sample. The RNA-Seq densities are aligned to the isoforms drawn at the bottom of the plot, which are automatically read from the GFF annotation of the events given as input.</li>
<li>Junction reads are visualized as arcs connecting the pair of exons that the junction borders. The thickness of the arc is in proportion to the number of junction reads present in the sample, but the actual number of junction reads can be optionally plotted too (as in the main example.)</li>
<li>MISO expression estimates are (optionally) shown on the right, including the full posterior distribution (as black histograms) over Ψ, with the Ψ estimate drawn as a thick red line and lower and upper 95% confidence intervals plotted as dotted grey lines. The actual value of Ψ along with the value of each confidence interval bound is shown to the right of the histograms.</li>
</ul>
</div>
</div>
<div class="section" id="overview">
<h1><a class="toc-backref" href="#id8">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<div class="section" id="visualizing-rna-seq-reads-along-isoforms-with-miso-estimates">
<span id="config"></span><h2><a class="toc-backref" href="#id9">Visualizing RNA-Seq reads along isoforms with MISO estimates</a><a class="headerlink" href="#visualizing-rna-seq-reads-along-isoforms-with-miso-estimates" title="Permalink to this headline">¶</a></h2>
<p>We return to our main test example of the <tt class="docutils literal"><span class="pre">--plot-event</span></tt> feature. The call:</p>
<div class="highlight-python"><pre>python plot.py --plot-event "chr17:45816186:45816265:-@chr17:45815912:45815950:-@chr17:45814875:45814965:-" test-data/event-data/ settings/sashimi_plot_settings.txt --output-dir test-plot</pre>
</div>
<p>Plots the event called <tt class="docutils literal"><span class="pre">chr17:45816186:45816265:-&#64;chr17:45815912:45815950:-&#64;chr17:45814875:45814965:-</span></tt>, using the directory pickled event <tt class="docutils literal"><span class="pre">test-data/event-data/</span></tt> and plotting according to the information provided in the settings file <tt class="docutils literal"><span class="pre">settings/sashimi_plot_settings.txt</span></tt>. The name of this event in this case is simply the ID given to this skipped exon in the GFF annotations provided with MISO (see <a class="reference external" href="http://genes.mit.edu/burgelab/miso/annotations/mm9_alt_events.zip">Mouse genome (mm9) alternative events</a>). The name is arbitrary, and <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> will visualize whatever events you give it as long as they have a corresponding indexed GFF file.</p>
<p>The directory containing the event/gene isoform information (in the above example, <tt class="docutils literal"><span class="pre">test-data/event-data</span></tt>) can be any directory generated by indexing a GFF3 file, using the <tt class="docutils literal"><span class="pre">index_gff.py</span></tt> script that is part of MISO. For more information on indexing, see <a class="reference internal" href="index.html#indexing-annotation"><em>Preparing the alternative isoforms annotation</em></a>.</p>
<p>The settings file for <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> specifies the name of each of the samples to be plotted, the directory containing their corresponding BAM files and MISO output, and a variety of plotting parameters, such as the figure colors and dimensions. The example settings file <tt class="docutils literal"><span class="pre">settings/sashimi_plot_settings.txt</span></tt> is:</p>
<div class="highlight-python"><pre>[data]
# directory where BAM files are
bam_prefix = ./test-data/bam-data/
# directory where MISO output is
miso_prefix = ./test-data/miso-data/

bam_files = [
    "heartWT1.sorted.bam",
    "heartWT2.sorted.bam",
    "heartKOa.sorted.bam",
    "heartKOb.sorted.bam"]

miso_files = [
    "heartWT1",
    "heartWT2",
    "heartKOa",
    "heartKOb"]

[plotting]
# Dimensions of figure to be plotted (in inches)
fig_width = 7
fig_height = 5
# Factor to scale down introns and exons by
intron_scale = 30
exon_scale = 4
# Whether to use a log scale or not when plotting
logged = False
font_size = 6

# Max y-axis
ymax = 150

# Whether to plot posterior distributions inferred by MISO
show_posteriors = True

# Whether to show posterior distributions as bar summaries
bar_posteriors = False

# Whether to plot the number of reads in each junction
number_junctions = True

resolution = .5
posterior_bins = 40
gene_posterior_ratio = 5

# List of colors for read denisites of each sample
colors = [
    "#CC0011",
    "#CC0011",
    "#FF8800",
    "#FF8800"]

# Number of mapped reads in each sample
# (Used to normalize the read density for RPKM calculation)
coverages = [
    6830944,
    14039751,
    4449737,
    6720151]

# Bar color for Bayes factor distribution
# plots (--plot-bf-dist)
# Paint them blue
bar_color = "b"

# Bayes factors thresholds to use for --plot-bf-dist
bf_thresholds = [0, 1, 2, 5, 10, 20]</pre>
</div>
<p>The above settings file specifies where the BAM files for each sample are (and their corresponding MISO output files) and also controls several useful plotting parameters. The parameters are:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">bam_prefix</span></tt>: directory where BAM files for the samples to plot are. These BAM files should be coordinate-sorted and indexed.</li>
<li><tt class="docutils literal"><span class="pre">miso_prefix</span></tt>: directory where MISO output directories are for the events to be plotted. For example, if plotting a skipped exon event for which the MISO output lives in <tt class="docutils literal"><span class="pre">/data/miso_output/SE/</span></tt>, then <tt class="docutils literal"><span class="pre">miso_prefix</span></tt> should be set to <tt class="docutils literal"><span class="pre">/data/miso_output/SE</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">bam_files</span></tt>: list of BAM files for RNA-Seq samples in the order in which you&#8217;d like them to be plotted. Each value in the list should be a filename that resides in the directory specified by <tt class="docutils literal"><span class="pre">bam_prefix</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">miso_files</span></tt>: list of MISO output directories for each sample. Should follow same order of samples as <tt class="docutils literal"><span class="pre">bam_files</span></tt>. Each value in the list should be a MISO output directory that resides in the directory specified by <tt class="docutils literal"><span class="pre">miso_prefix</span></tt>.</li>
</ul>
<blockquote>
<div><div class="admonition note">
<p class="first admonition-title">Note</p>
<p><tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> will look recursively in paths of <tt class="docutils literal"><span class="pre">miso_files</span></tt> to find the MISO output file (ending in <tt class="docutils literal"><span class="pre">.miso</span></tt>) associated with the event that is being plotted. For example, if we have these settings:</p>
<div class="highlight-python"><pre>miso_prefix = /miso/output/
miso_files = ['control']</pre>
</div>
<p class="last">If our event is on a chromosome called <tt class="docutils literal"><span class="pre">chr7</span></tt> in the annotation, then the program will check every subdirectory of <tt class="docutils literal"><span class="pre">/miso/output/control</span></tt> for a directory called <tt class="docutils literal"><span class="pre">chr7</span></tt>, and look for a file that has the form <tt class="docutils literal"><span class="pre">event_name.miso</span></tt> in that directory. If it cannot find such a directory in the first-level subdirectories, it will recurse into the subdirectories until it can find the file or until there are no more subdirectories to search.</p>
</div>
</div></blockquote>
</div></blockquote>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">fig_width</span></tt>: output figure&#8217;s width (in inches.)</li>
</ul>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">fig_height</span></tt>: output figure&#8217;s height (in inches.)</li>
<li><tt class="docutils literal"><span class="pre">exon_scale</span></tt> / <tt class="docutils literal"><span class="pre">intron_scale</span></tt>: factor by which to scale down exons and introns, respectively.</li>
<li><tt class="docutils literal"><span class="pre">logged</span></tt>: whether to log the RNA-Seq read densities (set to <tt class="xref docutils literal"><span class="pre">False</span></tt> for linear scaling)</li>
<li><tt class="docutils literal"><span class="pre">ymax</span></tt>: maximum value of y-axis for RNA-Seq read densities. If not given, then the highest y-axis value across all samples will be set for each, resulting in comparable y-scaling.</li>
<li><tt class="docutils literal"><span class="pre">show_posteriors</span></tt>: plot MISO posterior distributions if <tt class="xref docutils literal"><span class="pre">True</span></tt>, do not if <tt class="xref docutils literal"><span class="pre">False</span></tt></li>
<li><tt class="docutils literal"><span class="pre">bar_posteriors</span></tt>: plot MISO posterior distributions not as histograms, but as a horizontal bar that simply shows the mean and confidence intervals of the distribution in each sample.</li>
<li><tt class="docutils literal"><span class="pre">colors</span></tt>: Colors to use for each sample. Colors should be listed in same order as <tt class="docutils literal"><span class="pre">bam_files</span></tt> and <tt class="docutils literal"><span class="pre">miso_files</span></tt> lists.</li>
<li><tt class="docutils literal"><span class="pre">coverages</span></tt>: Number of mapping reads in each sample, for use when when computing normalized (i.e. RPKM) RNA-Seq read densities. Should be listed in same order as <tt class="docutils literal"><span class="pre">bam_files</span></tt> and <tt class="docutils literal"><span class="pre">miso_files</span></tt>. These numbers correspond to the &#8220;per million&#8221; denominators used for calculating RPKM.</li>
</ul>
</div></blockquote>
<p><strong>Additional parameters (all optional):</strong></p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">sample_labels</span></tt>: a list of string labels for each sample. By default, <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> will use the BAM filename from <tt class="docutils literal"><span class="pre">bam_files</span></tt> as the label for the sample. This option provides alternative labels. Note that <tt class="docutils literal"><span class="pre">sample_labels</span></tt> must have the same number of entries as <tt class="docutils literal"><span class="pre">bam_files</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">reverse_minus</span></tt>: specifies whether minus strand (<tt class="docutils literal"><span class="pre">-</span></tt>) event isoforms are to be plotted in same direction as plus strand events. By default, set to <tt class="xref docutils literal"><span class="pre">False</span></tt>, meaning minus strand events will be plotted in direction opposite to plus strand events.</li>
<li><tt class="docutils literal"><span class="pre">nxticks</span></tt>: number of x-axis ticks to plot</li>
<li><tt class="docutils literal"><span class="pre">nyticks</span></tt>: number of y-axis ticks to plot</li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For junction visualization, <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> currently uses only reads that cross <em>a single</em> junction. If a read crosses multiple exon-exon junctions, it is currently skipped, although MISO will use such a read in isoform estimation if it consistent with the given isoform annotation. Also, <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> currently ignores reads containing insertions or deletions and does not visualize sequence mismatches.</p>
</div>
</div>
<div class="section" id="command-line-options">
<h2><a class="toc-backref" href="#id10">Command-line options</a><a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">sashimi_plot/plot.py</span></tt> takes the following arguments:</p>
<div class="highlight-python"><pre>--plot-posterior
                      Plot the posterior distribution. Takes as input a raw
                      MISO output file (.miso)
--plot-insert-len
                      Plot the insert length distribution from a given
                      insert length (*.insert_len) filename. Second
                      argument is a settings filename.
--plot-bf-dist
                      Plot Bayes factor distributon. Takes the arguments:
                      (1) Bayes factor filename (*.miso_bf) settings
                      filename, (2) a settings filename.
--plot-event
                      Plot read densities and MISO inferences for a given
                      alternative event. Takes the arguments: (1) event name
                      (i.e. the ID= of the event based on MISO gff3
                      annotation file, (2) directory where MISO output is
                      for that event type (e.g. if event is a skipped exon,
                      provide the directory where the output for all SE
                      events are), (3) path to plotting settings file.
--output-dir
                      Output directory.</pre>
</div>
</div>
</div>
<div class="section" id="visualizing-and-plotting-miso-output">
<span id="plotting"></span><h1><a class="toc-backref" href="#id11">Visualizing and plotting MISO output</a><a class="headerlink" href="#visualizing-and-plotting-miso-output" title="Permalink to this headline">¶</a></h1>
<p>MISO comes with several built-in utilities for plotting its output, which all make use of the Python <tt class="docutils literal"><span class="pre">matplotlib</span></tt> environment package. These can be accessed through the <tt class="docutils literal"><span class="pre">plot.py</span></tt> utility.</p>
<div class="section" id="customizing-the-appearance-of-miso-estimates-alongside-raw-rna-seq-data">
<h2><a class="toc-backref" href="#id12">Customizing the appearance of MISO estimates alongside raw RNA-Seq data</a><a class="headerlink" href="#customizing-the-appearance-of-miso-estimates-alongside-raw-rna-seq-data" title="Permalink to this headline">¶</a></h2>
<p>In the main example of <tt class="docutils literal"><span class="pre">--plot-event</span></tt> shown above, the MISO posterior distributions are shown fully as a histogram. Sometimes it&#8217;s easier to compare a group of samples by just comparing the mean expression level (along with confidence intervals) in each sample, without plotting the entire distribution. Using the <tt class="docutils literal"><span class="pre">bar_posteriors</span></tt> option in the settings file, this can be done. Setting:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">bar_posteriors</span> <span class="o">=</span> <span class="bp">True</span>
</pre></div>
</div>
<p>yields the plot below:</p>
<a class="reference internal image-reference" href="_images/plot_event_bar_posteriors.png"><img alt="Showing summaries of MISO posterior distributions as horizontal bars" class="align-center" src="_images/plot_event_bar_posteriors.png" style="width: 808.8px; height: 561.6px;" /></a>
<p>The mean of each sample&#8217;s posterior distribution over Ψ is shown as a circle, with horizontal error bars extending to the upper and lower bounds of the confidence interval in each sample. Since the x-axis remains fixed in all samples, this makes it easy to visually compare the means of all samples and the overlap between their confidence intervals.</p>
</div>
<div class="section" id="plotting-the-distribution-of-events-with-bayes-factors">
<h2><a class="toc-backref" href="#id13">Plotting the distribution of events with Bayes factors</a><a class="headerlink" href="#plotting-the-distribution-of-events-with-bayes-factors" title="Permalink to this headline">¶</a></h2>
<p>It is often useful to plot the distribution of events that meet various Bayes factor thresholds. For any Bayes factor threshold, we can compute the number of events that meet that threshold in a given comparison file and visualize this as a distribution. The option <tt class="docutils literal"><span class="pre">--plot-bf-dist</span></tt> does this, as follows:</p>
<div class="highlight-python"><pre>plot.py --plot-bf-dist control_vs_knockdown.miso_bf settings.txt --output-dir plots/</pre>
</div>
<p>This will plot the distribution of events meeting various Bayes factors thresholds in the file <tt class="docutils literal"><span class="pre">control_vs_knockdown.miso_bf</span></tt> (outputted by calling <tt class="docutils literal"><span class="pre">--compare-samples</span></tt> in MISO) using the plotting settings file <tt class="docutils literal"><span class="pre">settings.txt</span></tt>, and output the resulting plot to <tt class="docutils literal"><span class="pre">plots/</span></tt>. The resulting plot will look like:</p>
<blockquote>
<div><div class="align-center figure align-center">
<a class="reference internal image-reference" href="_images/bf_dist.png"><img alt="Distribution of events meeting various Bayes factor thresholds" src="_images/bf_dist.png" style="width: 600.0px; height: 450.5px;" /></a>
<p class="caption"><em>Distribution of events meeting various Bayes factor thresholds</em></p>
</div>
</div></blockquote>
<p>This figure shows the number of events (in logarithmic scale) in the <tt class="docutils literal"><span class="pre">.miso_bf</span></tt> file that have Bayes factor greater than or equal to 0, greater than or equal to 1, greater than or equal to 2, etc. all the way to events with Bayes factor greater than or equal to 20.</p>
<p>The title of the plot says how many of the events in the input <tt class="docutils literal"><span class="pre">.miso_bf</span></tt> file were used in plotting the distribution. In the above example all 5231 entries in the file were used, but if the lowest Bayes factor threshold for the x-axis was set to be 2, for example, then only a subset of the entries would be plotted since there are events with Bayes factor less than 2.</p>
<p>The color of the bars used in the plot and Bayes factor thresholds for the x-axis can be customized through the setting file options <tt class="docutils literal"><span class="pre">bar_color</span></tt> and <tt class="docutils literal"><span class="pre">bf_thresholds</span></tt>, respectively. The default settings are:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># Color of bar for --plot-bf-dist</span>
<span class="n">bar_color</span> <span class="o">=</span> <span class="s">&quot;b&quot;</span>

<span class="c"># Bayes factor thresholds for --plot-bf-dist</span>
<span class="n">bf_thresholds</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="section" id="plotting-miso-estimates-individually">
<h2><a class="toc-backref" href="#id14">Plotting MISO estimates individually</a><a class="headerlink" href="#plotting-miso-estimates-individually" title="Permalink to this headline">¶</a></h2>
<p>We can visualize posterior distributions as histograms using the <tt class="docutils literal"><span class="pre">--plot-posterior</span></tt> argument to <tt class="docutils literal"><span class="pre">plot.py</span></tt>. For example, to plot the MISO posterior distribution of our example event in the <tt class="docutils literal"><span class="pre">heartKOa</span></tt> sample, run:</p>
<div class="highlight-python"><pre>python plot.py --plot-posterior "test-data/miso-data/heartKOa/chr17/chr17:45816186:45816265:-@chr17:45815912:45815950:-@chr17:45814875:45814965:-.miso" --output-dir test-plot</pre>
</div>
<p>This will produce a PDF plot that looks like this:</p>
<a class="reference internal image-reference" href="_images/posterior_plot1.png"><img alt="sashimi_plot of posterior distribution" class="align-center" src="_images/posterior_plot1.png" style="width: 525.5px; height: 410.0px;" /></a>
</div>
<div class="section" id="plotting-insert-length-distributions-and-summary-statistics">
<h2><a class="toc-backref" href="#id15">Plotting insert length distributions and summary statistics</a><a class="headerlink" href="#plotting-insert-length-distributions-and-summary-statistics" title="Permalink to this headline">¶</a></h2>
<p>For paired-end RNA-Seq samples, we can visualize the insert length distribution. This distribution is informative about the quality of the RNA-Seq sample, since it can tell us how precisely or cleanly the insert length of interest was selected during the RNA-Seq library preparation. This distribution is also used by <a class="reference external" href="http://genes.mit.edu/burgelab/miso/">MISO</a> in order to assign read pairs to isoforms, and so the tighter this distribution is, the more confident we can be in assigning read pairs to isoforms based on their insert length.</p>
<p>The distribution can be plotted using the <tt class="docutils literal"><span class="pre">--plot-insert-len</span></tt> option, which takes as input: (1) an insert length file (ending in <tt class="docutils literal"><span class="pre">.insert_len</span></tt>) produced by MISO and (2) a plotting settings filename. For example:</p>
<div class="highlight-python"><pre>python plot.py --plot-insert-len sample.insert_len settings.txt --output-dir plots/</pre>
</div>
<p>will produce a histogram of the insert length in <tt class="docutils literal"><span class="pre">sample.insert_len</span></tt> and place it in the <tt class="docutils literal"><span class="pre">plots</span></tt> directory. The histogram might look like this:</p>
<a class="reference internal image-reference" href="_images/insert-length-dist.png"><img alt="Insert length distribution plotted from a MISO insert length file" class="align-center" src="_images/insert-length-dist.png" style="width: 409.5px; height: 361.0px;" /></a>
</div>
</div>
<div class="section" id="frequently-asked-questions-faq">
<h1><a class="toc-backref" href="#id16">Frequently Asked Questions (FAQ)</a><a class="headerlink" href="#frequently-asked-questions-faq" title="Permalink to this headline">¶</a></h1>
<p>1. <strong>I&#8217;d like to plot RNA-Seq data for my own annotations, which are not part of the MISO events. Can this be done?</strong>
Yes. <tt class="docutils literal"><span class="pre">sashimi_plot</span></tt> can plot any event, as long as it is specified in the GFF3 format and indexed by the <tt class="docutils literal"><span class="pre">index_gff.py</span></tt> script that we provide. See <a class="reference internal" href="index.html#indexing-annotation"><em>Preparing the alternative isoforms annotation</em></a>.</p>
<p>2. <strong>I get the error that the</strong> <tt class="docutils literal"><span class="pre">.positions</span></tt> <strong>field is undefined.</strong>
This is caused by using an older version of the <a class="reference external" href="http://code.google.com/p/pysam/">pysam</a> module. Upgrading to version 0.6 or higher fixes the issue.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Section under construction</p>
</div>
</div>
<div class="section" id="authors">
<span id="katz"></span><span id="refs"></span><h1><a class="toc-backref" href="#id17">Authors</a><a class="headerlink" href="#authors" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li>Main feature (<tt class="docutils literal"><span class="pre">--plot-event</span></tt>) written by Eric T. Wang and Yarden Katz.</li>
<li>Other features written by Yarden Katz.</li>
</ul>
</div>
<div class="section" id="acknowledgements">
<h1><a class="toc-backref" href="#id18">Acknowledgements</a><a class="headerlink" href="#acknowledgements" title="Permalink to this headline">¶</a></h1>
<p>Thanks to:</p>
<ul class="simple">
<li>Vincent Butty (MIT)</li>
<li>Michael Lovci (UCSD)</li>
<li>Sol Katzman (UCSC)</li>
<li>Mohini Jangi (MIT)</li>
<li>Paul Boutz (MIT)</li>
<li>Sean O&#8217;Keeffe (Columbia)</li>
</ul>
</div>
<div class="section" id="references">
<h1><a class="toc-backref" href="#id19">References</a><a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h1>
<ol class="arabic simple">
<li>Katz Y, Wang ET, Airoldi EM, Burge CB. (2010). <a class="reference external" href="http://www.nature.com/nmeth/journal/v7/n12/full/nmeth.1528.html">Analysis and design of RNA sequencing experiments for identifying isoform regulation</a>. Nature Methods 7, 1009-1015.</li>
<li>Wang ET, Sandberg R, Luo S, Khrebtukova I, Zhang L, Mayr C, Kingsmore SF, Schroth GP, Burge CB. (2008). Alternative Isoform Regulation in Human Tissue Transcriptomes. Nature 456, 470-476</li>
</ol>
<div class="section" id="related-software">
<h2><a class="toc-backref" href="#id20">Related software</a><a class="headerlink" href="#related-software" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference external" href="http://www.broadinstitute.org/igv/">IGV</a>: Visualizer of mapped reads (e.g. BAM files). Displays junction reads.</li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">What is sashimi_plot?</a></li>
<li><a class="reference internal" href="#features">Features</a></li>
<li><a class="reference internal" href="#why-is-it-called-sashimi-plot">Why is it called sashimi_plot?</a></li>
<li><a class="reference internal" href="#updates">Updates</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#testing-sashimi-plot">Testing sashimi_plot</a></li>
</ul>
</li>
<li><a class="reference internal" href="#overview">Overview</a><ul>
<li><a class="reference internal" href="#visualizing-rna-seq-reads-along-isoforms-with-miso-estimates">Visualizing RNA-Seq reads along isoforms with MISO estimates</a></li>
<li><a class="reference internal" href="#command-line-options">Command-line options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#visualizing-and-plotting-miso-output">Visualizing and plotting MISO output</a><ul>
<li><a class="reference internal" href="#customizing-the-appearance-of-miso-estimates-alongside-raw-rna-seq-data">Customizing the appearance of MISO estimates alongside raw RNA-Seq data</a></li>
<li><a class="reference internal" href="#plotting-the-distribution-of-events-with-bayes-factors">Plotting the distribution of events with Bayes factors</a></li>
<li><a class="reference internal" href="#plotting-miso-estimates-individually">Plotting MISO estimates individually</a></li>
<li><a class="reference internal" href="#plotting-insert-length-distributions-and-summary-statistics">Plotting insert length distributions and summary statistics</a></li>
</ul>
</li>
<li><a class="reference internal" href="#frequently-asked-questions-faq">Frequently Asked Questions (FAQ)</a></li>
<li><a class="reference internal" href="#authors">Authors</a></li>
<li><a class="reference internal" href="#acknowledgements">Acknowledgements</a></li>
<li><a class="reference internal" href="#references">References</a><ul>
<li><a class="reference internal" href="#related-software">Related software</a></li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/sashimi.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li><a href="index.html">MISO v1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2010, Yarden Katz, Gábor Csárdi, Eric T. Wang, Edoardo M. Airoldi, Christopher B. Burge.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.7.
    </div>
  </body>
</html>